<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.3.1"/>
<title>Ensembl Genomes API: Bio::EnsEMBL::Utils::MetaData::GenomeInfo Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">Ensembl Genomes API
   </div>
   <div id="projectbrief">Auxillary Perl API for Ensembl Genomes</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.3.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>Bio</b></li><li class="navelem"><b>EnsEMBL</b></li><li class="navelem"><b>Utils</b></li><li class="navelem"><b>MetaData</b></li><li class="navelem"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">GenomeInfo</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Bio::EnsEMBL::Utils::MetaData::GenomeInfo Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a5547d9de15dcf3eaea35c6b37e09bc92"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a5547d9de15dcf3eaea35c6b37e09bc92">new</a> ()</td></tr>
<tr class="separator:a5547d9de15dcf3eaea35c6b37e09bc92"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac22dff02f6bb3bd3cce06c06c1afdc3d"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac22dff02f6bb3bd3cce06c06c1afdc3d">species</a> ()</td></tr>
<tr class="separator:ac22dff02f6bb3bd3cce06c06c1afdc3d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af078b69ba449d043cceae3bf3a69272e"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af078b69ba449d043cceae3bf3a69272e">strain</a> ()</td></tr>
<tr class="separator:af078b69ba449d043cceae3bf3a69272e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaadeee2f4c9733f317389fd57d9a08be"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aaadeee2f4c9733f317389fd57d9a08be">serotype</a> ()</td></tr>
<tr class="separator:aaadeee2f4c9733f317389fd57d9a08be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a101edd951fbee3f68259836cbb9b7af1"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a101edd951fbee3f68259836cbb9b7af1">name</a> ()</td></tr>
<tr class="separator:a101edd951fbee3f68259836cbb9b7af1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afa2ce749e440e2cc2809fb6c078c2ada"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afa2ce749e440e2cc2809fb6c078c2ada">dbname</a> ()</td></tr>
<tr class="separator:afa2ce749e440e2cc2809fb6c078c2ada"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4e4e9b068c87106f44d9bd6096b6edec"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4e4e9b068c87106f44d9bd6096b6edec">species_id</a> ()</td></tr>
<tr class="separator:a4e4e9b068c87106f44d9bd6096b6edec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a071717879973be03c76374af3f1e582a"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a071717879973be03c76374af3f1e582a">taxonomy_id</a> ()</td></tr>
<tr class="separator:a071717879973be03c76374af3f1e582a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac656cb0fb23ac490c3ce77731868c125"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac656cb0fb23ac490c3ce77731868c125">assembly_name</a> ()</td></tr>
<tr class="separator:ac656cb0fb23ac490c3ce77731868c125"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aceb2a9ccf2f2a968ce775125c9f93e17"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aceb2a9ccf2f2a968ce775125c9f93e17">assembly_id</a> ()</td></tr>
<tr class="separator:aceb2a9ccf2f2a968ce775125c9f93e17"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a26c408b9b27945cdeabc2a949d7e54dd"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a26c408b9b27945cdeabc2a949d7e54dd">assembly_level</a> ()</td></tr>
<tr class="separator:a26c408b9b27945cdeabc2a949d7e54dd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aba053334e9860264eb442e3e0cc8b3c9"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aba053334e9860264eb442e3e0cc8b3c9">genebuild</a> ()</td></tr>
<tr class="separator:aba053334e9860264eb442e3e0cc8b3c9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ace8f9eaee045d40cfc5ffffcc784f172"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ace8f9eaee045d40cfc5ffffcc784f172">division</a> ()</td></tr>
<tr class="separator:ace8f9eaee045d40cfc5ffffcc784f172"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a47833a4f3110976820e587c13d582cc5"><td class="memItemLeft" align="right" valign="top">public String&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a47833a4f3110976820e587c13d582cc5">db_size</a> ()</td></tr>
<tr class="separator:a47833a4f3110976820e587c13d582cc5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ade825d5953e67814db325554e7144d68"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ade825d5953e67814db325554e7144d68">base_count</a> ()</td></tr>
<tr class="separator:ade825d5953e67814db325554e7144d68"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a68ffa0a78969ff5ec0aa7caf0bef1eeb"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a68ffa0a78969ff5ec0aa7caf0bef1eeb">aliases</a> ()</td></tr>
<tr class="separator:a68ffa0a78969ff5ec0aa7caf0bef1eeb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac304c65827d50e84f81ada72e3d8bb1b"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac304c65827d50e84f81ada72e3d8bb1b">compara</a> ()</td></tr>
<tr class="separator:ac304c65827d50e84f81ada72e3d8bb1b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a537e48c3d95f8a42927fcff0d20faaa9"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a537e48c3d95f8a42927fcff0d20faaa9">sequences</a> ()</td></tr>
<tr class="separator:a537e48c3d95f8a42927fcff0d20faaa9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1cc2eddd90cfea6cdba23b8498192b2f"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a1cc2eddd90cfea6cdba23b8498192b2f">publications</a> ()</td></tr>
<tr class="separator:a1cc2eddd90cfea6cdba23b8498192b2f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a03cc3fba9fa7a7cd24ad172c45ac554b"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a03cc3fba9fa7a7cd24ad172c45ac554b">variations</a> ()</td></tr>
<tr class="separator:a03cc3fba9fa7a7cd24ad172c45ac554b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af0fd19e445a7a7d9d4f9fb6495152cfe"><td class="memItemLeft" align="right" valign="top">public Hashref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af0fd19e445a7a7d9d4f9fb6495152cfe">features</a> ()</td></tr>
<tr class="separator:af0fd19e445a7a7d9d4f9fb6495152cfe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a116215cff342a1ffb0f79d6b9168c692"><td class="memItemLeft" align="right" valign="top">public Hashref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a116215cff342a1ffb0f79d6b9168c692">annotations</a> ()</td></tr>
<tr class="separator:a116215cff342a1ffb0f79d6b9168c692"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9a6e4ac2019911b27e719d1c794de0c8"><td class="memItemLeft" align="right" valign="top">public Hashref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a9a6e4ac2019911b27e719d1c794de0c8">other_alignments</a> ()</td></tr>
<tr class="separator:a9a6e4ac2019911b27e719d1c794de0c8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a346b3ac487b2732b3f3a76bf7a657618"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a346b3ac487b2732b3f3a76bf7a657618">has_variations</a> ()</td></tr>
<tr class="separator:a346b3ac487b2732b3f3a76bf7a657618"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab4fdd35cf8eec7a5592600b8c041e007"><td class="memItemLeft" align="right" valign="top">public Boolean&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ab4fdd35cf8eec7a5592600b8c041e007">has_genome_alignments</a> ()</td></tr>
<tr class="separator:ab4fdd35cf8eec7a5592600b8c041e007"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a46b09f24242548113c754f6d73c932ab"><td class="memItemLeft" align="right" valign="top">public Boolean&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a46b09f24242548113c754f6d73c932ab">has_synteny</a> ()</td></tr>
<tr class="separator:a46b09f24242548113c754f6d73c932ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a40befb3b0bb11a8338962269fdb7f475"><td class="memItemLeft" align="right" valign="top">public Boolean&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a40befb3b0bb11a8338962269fdb7f475">has_peptide_compara</a> ()</td></tr>
<tr class="separator:a40befb3b0bb11a8338962269fdb7f475"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4de11c6d56a4d9229f6cdd1a44ababf7"><td class="memItemLeft" align="right" valign="top">public Boolean&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4de11c6d56a4d9229f6cdd1a44ababf7">has_pan_compara</a> ()</td></tr>
<tr class="separator:a4de11c6d56a4d9229f6cdd1a44ababf7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aae89a27e1bf3e2b99f5ed03481d1c5dd"><td class="memItemLeft" align="right" valign="top">public Boolean&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aae89a27e1bf3e2b99f5ed03481d1c5dd">has_other_alignments</a> ()</td></tr>
<tr class="separator:aae89a27e1bf3e2b99f5ed03481d1c5dd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2fdd5e1fd00dc66e2d006cc3b9ffe2c8"><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a2fdd5e1fd00dc66e2d006cc3b9ffe2c8">count_variation</a> ()</td></tr>
<tr class="separator:a2fdd5e1fd00dc66e2d006cc3b9ffe2c8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8722a03a4806c7aed450fead382da626"><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a8722a03a4806c7aed450fead382da626">count_alignments</a> ()</td></tr>
<tr class="separator:a8722a03a4806c7aed450fead382da626"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a97b834a2ea2aff86ffa38ae95a311e29"><td class="memItemLeft" align="right" valign="top">public Uniprot&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a97b834a2ea2aff86ffa38ae95a311e29">get_uniprot_coverage</a> ()</td></tr>
<tr class="separator:a97b834a2ea2aff86ffa38ae95a311e29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa88bd8a7e5fb478fbd77d1d1db3ce841"><td class="memItemLeft" align="right" valign="top">public Hashref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aa88bd8a7e5fb478fbd77d1d1db3ce841">to_hash</a> ()</td></tr>
<tr class="separator:aa88bd8a7e5fb478fbd77d1d1db3ce841"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afa98af8afcbed837673491400a0196f7"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afa98af8afcbed837673491400a0196f7">count_hash_values</a> ()</td></tr>
<tr class="separator:afa98af8afcbed837673491400a0196f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a21af8b6060ecefc8cffa65a1bf63d93a"><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a21af8b6060ecefc8cffa65a1bf63d93a">count_hash_lengths</a> ()</td></tr>
<tr class="separator:a21af8b6060ecefc8cffa65a1bf63d93a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aea28c56ad95ba81c3a753fc3b242dea6"><td class="memItemLeft" align="right" valign="top">public DbID&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aea28c56ad95ba81c3a753fc3b242dea6">dbID</a> ()</td></tr>
<tr class="separator:aea28c56ad95ba81c3a753fc3b242dea6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae5ce14c54fd43a67421370bb931de12b"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ae5ce14c54fd43a67421370bb931de12b">adaptor</a> ()</td></tr>
<tr class="separator:ae5ce14c54fd43a67421370bb931de12b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afc32ee07463b73587c0d7fa6ac024cb3"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afc32ee07463b73587c0d7fa6ac024cb3">_preload</a> ()</td></tr>
<tr class="separator:afc32ee07463b73587c0d7fa6ac024cb3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a18e8f968f1ee3d82c6122ed60a8a0fbc"><td class="memItemLeft" align="right" valign="top">protected&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a18e8f968f1ee3d82c6122ed60a8a0fbc">_unload</a> ()</td></tr>
<tr class="separator:a18e8f968f1ee3d82c6122ed60a8a0fbc"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h1><a class="anchor" id="Synopsis"></a>
Synopsis</h1>
<div class="fragment"><div class="line">my $genome = <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a5547d9de15dcf3eaea35c6b37e09bc92">new</a>(</div>
<div class="line">    -<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac22dff02f6bb3bd3cce06c06c1afdc3d">species</a>    =&gt; $dba-&gt;species(),</div>
<div class="line">    -<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4e4e9b068c87106f44d9bd6096b6edec">species_id</a> =&gt; $dba-&gt;species_id(),</div>
<div class="line">    -<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ace8f9eaee045d40cfc5ffffcc784f172">division</a>   =&gt; $meta-&gt;get_division() || <span class="stringliteral">&#39;Ensembl&#39;</span>,</div>
<div class="line">    -<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afa2ce749e440e2cc2809fb6c078c2ada">dbname</a>     =&gt; $dbname);</div>
<div class="line">    </div>
<div class="line">print Dumper($genome-&gt;to_hash());</div>
</div><!-- fragment --> <h1><a class="anchor" id="Description"></a>
Description</h1>
 <pre>
Object encapsulating meta information about a genome in Ensembl Genomes. 

Can be used to render information about a genome e.g.

print $genome->name()." (".$genome->species.")\n";
print "Sequences: ".scalar(\@{$genome->sequences()})."\n";
if($genome->has_variations()) {
    print "Variations: \n";
    # variations is a hash with type as the key
    while(my ($type,$value) = each %{$genome->variations()}) {
        print "- $type\n";
    }
}
print "Compara analyses: ".scalar(\@{$genome->compara()})."\n";

</pre> </div><h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="afc32ee07463b73587c0d7fa6ac024cb3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::GenomeInfo::_preload </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Description: Ensure all children are loaded (used for hash transformation)
  Returntype : none
  Exceptions : none
  Caller     : Internal
  Status     : Stable
 </pre><div id="codesection-_preload" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_preload-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_preload-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_preload-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afc32ee07463b73587c0d7fa6ac024cb3">_preload</a> {</div>
<div class="line">  my ($self) = @_;</div>
<div class="line">  $self-&gt;aliases();</div>
<div class="line">  $self-&gt;annotations();</div>
<div class="line">  $self-&gt;compara();</div>
<div class="line">  $self-&gt;features();</div>
<div class="line">  $self-&gt;other_alignments();</div>
<div class="line">  $self-&gt;publications();</div>
<div class="line">  $self-&gt;sequences();</div>
<div class="line">  $self-&gt;variations();</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div><pre>  Description: Remove all children (used after hash transformation to ensure object is minimised)
  Returntype : none
  Exceptions : none
  Caller     : dump_metadata.pl
  Status     : Stable
 </pre><p> /** </p>

</div>
</div>
<a class="anchor" id="a18e8f968f1ee3d82c6122ed60a8a0fbc"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected Bio::EnsEMBL::Utils::MetaData::GenomeInfo::_unload </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-_unload" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_unload-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_unload-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_unload-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a18e8f968f1ee3d82c6122ed60a8a0fbc">_unload</a> {</div>
<div class="line">  my ($self) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a68ffa0a78969ff5ec0aa7caf0bef1eeb">aliases</a>}          = undef;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a116215cff342a1ffb0f79d6b9168c692">annotations</a>}      = undef;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac304c65827d50e84f81ada72e3d8bb1b">compara</a>}          = undef;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af0fd19e445a7a7d9d4f9fb6495152cfe">features</a>}         = undef;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a9a6e4ac2019911b27e719d1c794de0c8">other_alignments</a>} = undef;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a1cc2eddd90cfea6cdba23b8498192b2f">publications</a>}     = undef;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a537e48c3d95f8a42927fcff0d20faaa9">sequences</a>}        = undef;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a03cc3fba9fa7a7cd24ad172c45ac554b">variations</a>}       = undef;</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ae5ce14c54fd43a67421370bb931de12b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor</a> Bio::EnsEMBL::Utils::MetaData::GenomeInfo::adaptor </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) adaptor to set set
  Description: Gets/sets GenomeInfoAdaptor
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor</a>
  Exceptions : none
  Caller     : Internal
  Status     : Stable
 </pre><div id="codesection-adaptor" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-adaptor-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-adaptor-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-adaptor-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ae5ce14c54fd43a67421370bb931de12b">adaptor</a> {</div>
<div class="line">  my ( $self, $adaptor ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ae5ce14c54fd43a67421370bb931de12b">adaptor</a>} = $adaptor <span class="keywordflow">if</span> ( defined $adaptor );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ae5ce14c54fd43a67421370bb931de12b">adaptor</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a68ffa0a78969ff5ec0aa7caf0bef1eeb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::GenomeInfo::aliases </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) arrayref of aliases to set
  Description: Gets/sets aliases by which the genome is also known 
  Returntype : Arrayref of aliases
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-aliases" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-aliases-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-aliases-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-aliases-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a68ffa0a78969ff5ec0aa7caf0bef1eeb">aliases</a> {</div>
<div class="line">  my ( $self, $aliases ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $aliases ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a68ffa0a78969ff5ec0aa7caf0bef1eeb">aliases</a>} = $aliases;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a68ffa0a78969ff5ec0aa7caf0bef1eeb">aliases</a>} &amp;&amp; defined $self-&gt;adaptor() ) {</div>
<div class="line">    $self-&gt;adaptor()-&gt;_fetch_aliases($self);</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a68ffa0a78969ff5ec0aa7caf0bef1eeb">aliases</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a116215cff342a1ffb0f79d6b9168c692"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hashref Bio::EnsEMBL::Utils::MetaData::GenomeInfo::annotations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) annotations to set
  Description: Gets/sets summary information about gene annotation as hashref, with
               annotation type as key and count as value
  Returntype : hashref
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-annotations" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-annotations-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-annotations-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-annotations-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a116215cff342a1ffb0f79d6b9168c692">annotations</a> {</div>
<div class="line">  my ( $self, $annotation ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $annotation ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a116215cff342a1ffb0f79d6b9168c692">annotations</a>} = $annotation;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a116215cff342a1ffb0f79d6b9168c692">annotations</a>} &amp;&amp; defined $self-&gt;adaptor() ) {</div>
<div class="line">    $self-&gt;adaptor()-&gt;_fetch_annotations($self);</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a116215cff342a1ffb0f79d6b9168c692">annotations</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aceb2a9ccf2f2a968ce775125c9f93e17"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::assembly_id </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) assembly_id to set
  Description: Gets/sets INSDC accession for assembly
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-assembly_id" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-assembly_id-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-assembly_id-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-assembly_id-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aceb2a9ccf2f2a968ce775125c9f93e17">assembly_id</a> {</div>
<div class="line">  my ( $self, $assembly_id ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aceb2a9ccf2f2a968ce775125c9f93e17">assembly_id</a>} = $assembly_id <span class="keywordflow">if</span> ( defined $assembly_id );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aceb2a9ccf2f2a968ce775125c9f93e17">assembly_id</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a26c408b9b27945cdeabc2a949d7e54dd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::assembly_level </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) assembly_level to set
  Description: Gets/sets highest level of assembly (chromosome, supercontig etc.)
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-assembly_level" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-assembly_level-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-assembly_level-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-assembly_level-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a26c408b9b27945cdeabc2a949d7e54dd">assembly_level</a> {</div>
<div class="line">  my ( $self, $assembly_level ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a26c408b9b27945cdeabc2a949d7e54dd">assembly_level</a>} = $assembly_level</div>
<div class="line">    <span class="keywordflow">if</span> ( defined $assembly_level );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a26c408b9b27945cdeabc2a949d7e54dd">assembly_level</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac656cb0fb23ac490c3ce77731868c125"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::assembly_name </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) assembly_name to set
  Description: Gets/sets name of assembly
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-assembly_name" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-assembly_name-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-assembly_name-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-assembly_name-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac656cb0fb23ac490c3ce77731868c125">assembly_name</a> {</div>
<div class="line">  my ( $self, $assembly_name ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac656cb0fb23ac490c3ce77731868c125">assembly_name</a>} = $assembly_name <span class="keywordflow">if</span> ( defined $assembly_name );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac656cb0fb23ac490c3ce77731868c125">assembly_name</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ade825d5953e67814db325554e7144d68"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::GenomeInfo::base_count </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-base_count" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-base_count-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-base_count-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-base_count-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ade825d5953e67814db325554e7144d68">base_count</a> {</div>
<div class="line">  my ( $self, $base_count ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ade825d5953e67814db325554e7144d68">base_count</a>} = $base_count <span class="keywordflow">if</span> ( defined $base_count );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ade825d5953e67814db325554e7144d68">base_count</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac304c65827d50e84f81ada72e3d8bb1b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::GenomeInfo::compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">GenomeComparaInfo</a> objects to set
  Description: Gets/sets <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">GenomeComparaInfo</a> describing comparative analyses applied to the genome
  Returntype : Arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac304c65827d50e84f81ada72e3d8bb1b">compara</a> {</div>
<div class="line">  my ( $self, $compara ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $compara ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac304c65827d50e84f81ada72e3d8bb1b">compara</a>}               = $compara;</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a40befb3b0bb11a8338962269fdb7f475">has_peptide_compara</a>}   = undef;</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a46b09f24242548113c754f6d73c932ab">has_synteny</a>}           = undef;</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ab4fdd35cf8eec7a5592600b8c041e007">has_genome_alignments</a>} = undef;</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4de11c6d56a4d9229f6cdd1a44ababf7">has_pan_compara</a>}       = undef;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac304c65827d50e84f81ada72e3d8bb1b">compara</a>} &amp;&amp; defined $self-&gt;adaptor() ) {</div>
<div class="line">    $self-&gt;adaptor()-&gt;_fetch_comparas($self);</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac304c65827d50e84f81ada72e3d8bb1b">compara</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a8722a03a4806c7aed450fead382da626"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::Utils::MetaData::GenomeInfo::count_alignments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Description: Returns total number of alignments to genome
  Returntype : integer
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-count_alignments" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-count_alignments-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_alignments-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_alignments-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a8722a03a4806c7aed450fead382da626">count_alignments</a> {</div>
<div class="line">  my ($self) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;count_hash_values( $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a9a6e4ac2019911b27e719d1c794de0c8">other_alignments</a>}{bam} ) +</div>
<div class="line">    $self-&gt;count_hash_values(</div>
<div class="line">                     $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a9a6e4ac2019911b27e719d1c794de0c8">other_alignments</a>}{proteinAlignFeatures} ) +</div>
<div class="line">    $self-&gt;count_hash_values(</div>
<div class="line">                          $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a9a6e4ac2019911b27e719d1c794de0c8">other_alignments</a>}{dnaAlignFeatures} );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a21af8b6060ecefc8cffa65a1bf63d93a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::Utils::MetaData::GenomeInfo::count_hash_lengths </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Description: Sums sizes of arrays found in hash as values
  Arg        : hashref
  Returntype : integer
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_hash_lengths" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-count_hash_lengths-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_hash_lengths-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_hash_lengths-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a21af8b6060ecefc8cffa65a1bf63d93a">count_hash_lengths</a> {</div>
<div class="line">  my ( $self, $hash ) = @_;</div>
<div class="line">  my $tot = 0;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $hash ) {</div>
<div class="line">    <span class="keywordflow">for</span> my $v ( values %{$hash} ) {</div>
<div class="line">      $tot += scalar(@$v);</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $tot;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="afa98af8afcbed837673491400a0196f7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::GenomeInfo::count_hash_values </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-count_hash_values" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-count_hash_values-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_hash_values-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_hash_values-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afa98af8afcbed837673491400a0196f7">count_hash_values</a> {</div>
<div class="line">  my ( $self, $hash ) = @_;</div>
<div class="line">  my $tot = 0;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $hash ) {</div>
<div class="line">    <span class="keywordflow">for</span> my $v ( values %{$hash} ) {</div>
<div class="line">      $tot += $v;</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $tot;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a2fdd5e1fd00dc66e2d006cc3b9ffe2c8"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::Utils::MetaData::GenomeInfo::count_variation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Description: Returns total number of variations and structural variations mapped to genome
  Returntype : integer
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-count_variation" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-count_variation-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_variation-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_variation-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a2fdd5e1fd00dc66e2d006cc3b9ffe2c8">count_variation</a> {</div>
<div class="line">  my ($self) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;count_hash_values( $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a03cc3fba9fa7a7cd24ad172c45ac554b">variations</a>}{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a03cc3fba9fa7a7cd24ad172c45ac554b">variations</a>} ) +</div>
<div class="line">    $self-&gt;count_hash_values(</div>
<div class="line">                           $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a03cc3fba9fa7a7cd24ad172c45ac554b">variations</a>}{structural_variations} );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a47833a4f3110976820e587c13d582cc5"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::db_size </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) db_size to set
  Description: Gets/sets size of database containing core
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-db_size" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-db_size-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-db_size-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-db_size-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a47833a4f3110976820e587c13d582cc5">db_size</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a47833a4f3110976820e587c13d582cc5">db_size</a>} = $arg <span class="keywordflow">if</span> ( defined $arg );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a47833a4f3110976820e587c13d582cc5">db_size</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aea28c56ad95ba81c3a753fc3b242dea6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public DbID Bio::EnsEMBL::Utils::MetaData::GenomeInfo::dbID </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) base_count to set
  Description: Gets/sets total number of bases in assembled genome
  Returntype : integer
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><p> /**</p>
<pre>  Arg        : (optional) dbID to set set
  Description: Gets/sets internal genome_id used as database primary key
  Returntype : dbID string
  Exceptions : none
  Caller     : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor</a>
  Status     : Stable
 </pre><div id="codesection-dbID" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-dbID-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-dbID-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-dbID-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aea28c56ad95ba81c3a753fc3b242dea6">dbID</a> {</div>
<div class="line">  my ( $self, $id ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aea28c56ad95ba81c3a753fc3b242dea6">dbID</a>} = $id <span class="keywordflow">if</span> ( defined $id );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aea28c56ad95ba81c3a753fc3b242dea6">dbID</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="afa2ce749e440e2cc2809fb6c078c2ada"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::dbname </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) dbname to set
  Description: Gets/sets name of core database from which genome comes
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-dbname" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-dbname-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-dbname-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-dbname-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afa2ce749e440e2cc2809fb6c078c2ada">dbname</a> {</div>
<div class="line">  my ( $self, $dbname ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afa2ce749e440e2cc2809fb6c078c2ada">dbname</a>} = $dbname <span class="keywordflow">if</span> ( defined $dbname );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afa2ce749e440e2cc2809fb6c078c2ada">dbname</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ace8f9eaee045d40cfc5ffffcc784f172"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::division </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) division to set
  Description: Gets/sets Ensembl Genomes division
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-division" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-division-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-division-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-division-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ace8f9eaee045d40cfc5ffffcc784f172">division</a> {</div>
<div class="line">  my ( $self, $division ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ace8f9eaee045d40cfc5ffffcc784f172">division</a>} = $division <span class="keywordflow">if</span> ( defined $division );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ace8f9eaee045d40cfc5ffffcc784f172">division</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="af0fd19e445a7a7d9d4f9fb6495152cfe"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hashref Bio::EnsEMBL::Utils::MetaData::GenomeInfo::features </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) features to set
  Description: Gets/sets general genomic features associated with the genome as hashref
               keyed by type (e.g. repeatFeatures,simpleFeatures), further broken down into
               counts by analysis
  Returntype : Hashref
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-features" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-features-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-features-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-features-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af0fd19e445a7a7d9d4f9fb6495152cfe">features</a> {</div>
<div class="line">  my ( $self, $features ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $features ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af0fd19e445a7a7d9d4f9fb6495152cfe">features</a>} = $features;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af0fd19e445a7a7d9d4f9fb6495152cfe">features</a>} &amp;&amp; defined $self-&gt;adaptor() ) {</div>
<div class="line">    $self-&gt;adaptor()-&gt;_fetch_features($self);</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af0fd19e445a7a7d9d4f9fb6495152cfe">features</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aba053334e9860264eb442e3e0cc8b3c9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::genebuild </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) genebuild to set
  Description: Gets/sets identifier for genebuild
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-genebuild" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-genebuild-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-genebuild-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-genebuild-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aba053334e9860264eb442e3e0cc8b3c9">genebuild</a> {</div>
<div class="line">  my ( $self, $genebuild ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aba053334e9860264eb442e3e0cc8b3c9">genebuild</a>} = $genebuild <span class="keywordflow">if</span> ( defined $genebuild );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aba053334e9860264eb442e3e0cc8b3c9">genebuild</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a97b834a2ea2aff86ffa38ae95a311e29"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Uniprot Bio::EnsEMBL::Utils::MetaData::GenomeInfo::get_uniprot_coverage </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Description: Get % of protein coding genes with a UniProt cross-reference
  Returntype : uniprot coverage as percentage
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-get_uniprot_coverage" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-get_uniprot_coverage-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-get_uniprot_coverage-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-get_uniprot_coverage-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a97b834a2ea2aff86ffa38ae95a311e29">get_uniprot_coverage</a> {</div>
<div class="line">  my ($self) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> 100.0*</div>
<div class="line">    ( $self-&gt;annotations()-&gt;{nProteinCodingUniProtKB} )/</div>
<div class="line">    $self-&gt;annotations()-&gt;{nProteinCoding};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ab4fdd35cf8eec7a5592600b8c041e007"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Boolean Bio::EnsEMBL::Utils::MetaData::GenomeInfo::has_genome_alignments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) 1/0 to set if genome has genome alignments
  Description: Boolean-style method, returns 1 if genome has genome alignments, 0 if not
  Returntype : 1 or 0
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-has_genome_alignments" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-has_genome_alignments-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-has_genome_alignments-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-has_genome_alignments-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ab4fdd35cf8eec7a5592600b8c041e007">has_genome_alignments</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $arg ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ab4fdd35cf8eec7a5592600b8c041e007">has_genome_alignments</a>} = $arg;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined( $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ab4fdd35cf8eec7a5592600b8c041e007">has_genome_alignments</a>} ) &amp;&amp;</div>
<div class="line">          defined $self-&gt;compara() )</div>
<div class="line">  {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ab4fdd35cf8eec7a5592600b8c041e007">has_genome_alignments</a>} = 0;</div>
<div class="line">    <span class="keywordflow">for</span> my $compara ( @{ $self-&gt;compara() } ) {</div>
<div class="line">      <span class="keywordflow">if</span> ( $compara-&gt;is_dna_compara() ) {</div>
<div class="line">        $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ab4fdd35cf8eec7a5592600b8c041e007">has_genome_alignments</a>} = 1;</div>
<div class="line">        last;</div>
<div class="line">      }</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ab4fdd35cf8eec7a5592600b8c041e007">has_genome_alignments</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aae89a27e1bf3e2b99f5ed03481d1c5dd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Boolean Bio::EnsEMBL::Utils::MetaData::GenomeInfo::has_other_alignments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) 1/0 to set if genome has other alignments
  Description: Boolean-style method, returns 1 if genome has other alignments, 0 if not
  Returntype : 1 or 0
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-has_other_alignments" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-has_other_alignments-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-has_other_alignments-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-has_other_alignments-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aae89a27e1bf3e2b99f5ed03481d1c5dd">has_other_alignments</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $arg ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aae89a27e1bf3e2b99f5ed03481d1c5dd">has_other_alignments</a>} = $arg;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined( $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aae89a27e1bf3e2b99f5ed03481d1c5dd">has_other_alignments</a>} ) ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aae89a27e1bf3e2b99f5ed03481d1c5dd">has_other_alignments</a>} =</div>
<div class="line">      $self-&gt;count_alignments() &gt; 0 ? 1 : 0;</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aae89a27e1bf3e2b99f5ed03481d1c5dd">has_other_alignments</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a4de11c6d56a4d9229f6cdd1a44ababf7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Boolean Bio::EnsEMBL::Utils::MetaData::GenomeInfo::has_pan_compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) 1/0 to set if genome is included in pan compara
  Description: Boolean-style method, returns 1 if genome is in pan compara, 0 if not
  Returntype : 1 or 0
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-has_pan_compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-has_pan_compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-has_pan_compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-has_pan_compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4de11c6d56a4d9229f6cdd1a44ababf7">has_pan_compara</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $arg ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4de11c6d56a4d9229f6cdd1a44ababf7">has_pan_compara</a>} = $arg;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined( $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4de11c6d56a4d9229f6cdd1a44ababf7">has_pan_compara</a>} ) &amp;&amp;</div>
<div class="line">          defined $self-&gt;compara() )</div>
<div class="line">  {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4de11c6d56a4d9229f6cdd1a44ababf7">has_pan_compara</a>} = 0;</div>
<div class="line">    <span class="keywordflow">for</span> my $compara ( @{ $self-&gt;compara() } ) {</div>
<div class="line">      <span class="keywordflow">if</span> ( $compara-&gt;is_pan_compara() ) {</div>
<div class="line">        $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4de11c6d56a4d9229f6cdd1a44ababf7">has_pan_compara</a>} = 1;</div>
<div class="line">        last;</div>
<div class="line">      }</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4de11c6d56a4d9229f6cdd1a44ababf7">has_pan_compara</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a40befb3b0bb11a8338962269fdb7f475"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Boolean Bio::EnsEMBL::Utils::MetaData::GenomeInfo::has_peptide_compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) 1/0 to set if genome has peptide compara
  Description: Boolean-style method, returns 1 if genome has peptide, 0 if not
  Returntype : 1 or 0
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-has_peptide_compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-has_peptide_compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-has_peptide_compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-has_peptide_compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a40befb3b0bb11a8338962269fdb7f475">has_peptide_compara</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $arg ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a40befb3b0bb11a8338962269fdb7f475">has_peptide_compara</a>} = $arg;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined( $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a40befb3b0bb11a8338962269fdb7f475">has_peptide_compara</a>} ) &amp;&amp;</div>
<div class="line">          defined $self-&gt;compara() )</div>
<div class="line">  {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a40befb3b0bb11a8338962269fdb7f475">has_peptide_compara</a>} = 0;</div>
<div class="line">    <span class="keywordflow">for</span> my $compara ( @{ $self-&gt;compara() } ) {</div>
<div class="line">      <span class="keywordflow">if</span> ( $compara-&gt;is_peptide_compara() &amp;&amp; !$compara-&gt;is_pan_compara() ) {</div>
<div class="line">        $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a40befb3b0bb11a8338962269fdb7f475">has_peptide_compara</a>} = 1;</div>
<div class="line">        last;</div>
<div class="line">      }</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a40befb3b0bb11a8338962269fdb7f475">has_peptide_compara</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a46b09f24242548113c754f6d73c932ab"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Boolean Bio::EnsEMBL::Utils::MetaData::GenomeInfo::has_synteny </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) 1/0 to set if genome has synteny
  Description: Boolean-style method, returns 1 if genome has synteny, 0 if not
  Returntype : 1 or 0
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-has_synteny" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-has_synteny-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-has_synteny-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-has_synteny-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a46b09f24242548113c754f6d73c932ab">has_synteny</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $arg ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a46b09f24242548113c754f6d73c932ab">has_synteny</a>} = $arg;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined( $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a46b09f24242548113c754f6d73c932ab">has_synteny</a>} ) &amp;&amp; defined $self-&gt;compara() )</div>
<div class="line">  {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a46b09f24242548113c754f6d73c932ab">has_synteny</a>} = 0;</div>
<div class="line">    <span class="keywordflow">for</span> my $compara ( @{ $self-&gt;compara() } ) {</div>
<div class="line">      <span class="keywordflow">if</span> ( $compara-&gt;is_synteny() ) {</div>
<div class="line">        $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a46b09f24242548113c754f6d73c932ab">has_synteny</a>} = 1;</div>
<div class="line">        last;</div>
<div class="line">      }</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a46b09f24242548113c754f6d73c932ab">has_synteny</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a346b3ac487b2732b3f3a76bf7a657618"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::GenomeInfo::has_variations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-has_variations" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-has_variations-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-has_variations-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-has_variations-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a346b3ac487b2732b3f3a76bf7a657618">has_variations</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $arg ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a346b3ac487b2732b3f3a76bf7a657618">has_variations</a>} = $arg;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined( $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a346b3ac487b2732b3f3a76bf7a657618">has_variations</a>} ) ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a346b3ac487b2732b3f3a76bf7a657618">has_variations</a>} = $self-&gt;count_variation() &gt; 0 ? 1 : 0;</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a346b3ac487b2732b3f3a76bf7a657618">has_variations</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a101edd951fbee3f68259836cbb9b7af1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::name </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) name to set
  Description: Gets/sets readable display name for genome
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-name" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-name-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-name-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-name-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a101edd951fbee3f68259836cbb9b7af1">name</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a101edd951fbee3f68259836cbb9b7af1">name</a>} = $arg <span class="keywordflow">if</span> ( defined $arg );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a101edd951fbee3f68259836cbb9b7af1">name</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a5547d9de15dcf3eaea35c6b37e09bc92"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::GenomeInfo::new </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-new" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-new-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-new-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-new-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <span class="keyword">new</span> {</div>
<div class="line">  my ( $proto, @args ) = @_;</div>
<div class="line">  my $class = ref($proto) || $proto;</div>
<div class="line">  my $self = bless( {}, $class );</div>
<div class="line">  ( $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a101edd951fbee3f68259836cbb9b7af1">name</a>},       $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac22dff02f6bb3bd3cce06c06c1afdc3d">species</a>},     $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#afa2ce749e440e2cc2809fb6c078c2ada">dbname</a>},</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4e4e9b068c87106f44d9bd6096b6edec">species_id</a>}, $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a071717879973be03c76374af3f1e582a">taxonomy_id</a>}, $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac656cb0fb23ac490c3ce77731868c125">assembly_name</a>},</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aceb2a9ccf2f2a968ce775125c9f93e17">assembly_id</a>}, $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a26c408b9b27945cdeabc2a949d7e54dd">assembly_level</a>}, $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aba053334e9860264eb442e3e0cc8b3c9">genebuild</a>},</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ace8f9eaee045d40cfc5ffffcc784f172">division</a>},    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af078b69ba449d043cceae3bf3a69272e">strain</a>},         $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aaadeee2f4c9733f317389fd57d9a08be">serotype</a>}</div>
<div class="line"></div>
<div class="line">    ) =</div>
<div class="line">    rearrange(</div>
<div class="line">               [ <span class="stringliteral">&#39;NAME&#39;</span>,        <span class="stringliteral">&#39;SPECIES&#39;</span>,</div>
<div class="line">                 <span class="stringliteral">&#39;DBNAME&#39;</span>,      <span class="stringliteral">&#39;SPECIES_ID&#39;</span>,</div>
<div class="line">                 <span class="stringliteral">&#39;TAXONOMY_ID&#39;</span>, <span class="stringliteral">&#39;ASSEMBLY_NAME&#39;</span>,</div>
<div class="line">                 <span class="stringliteral">&#39;ASSEMBLY_ID&#39;</span>, <span class="stringliteral">&#39;ASSEMBLY_LEVEL&#39;</span>,</div>
<div class="line">                 <span class="stringliteral">&#39;GENEBUILD&#39;</span>,   <span class="stringliteral">&#39;DIVISION&#39;</span>,</div>
<div class="line">                 <span class="stringliteral">&#39;STRAIN&#39;</span>,      <span class="stringliteral">&#39;SEROTYPE&#39;</span> ],</div>
<div class="line">               @args );</div>
<div class="line">  <span class="keywordflow">return</span> $self;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a9a6e4ac2019911b27e719d1c794de0c8"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hashref Bio::EnsEMBL::Utils::MetaData::GenomeInfo::other_alignments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) other alignments to set
  Description: Gets/sets other alignments as hashref, keyed by type (dnaAlignFeatures,proteinAlignFeatures)
               with values as logic_name-count pairs 
  Returntype : Hashref
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-other_alignments" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-other_alignments-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-other_alignments-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-other_alignments-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a9a6e4ac2019911b27e719d1c794de0c8">other_alignments</a> {</div>
<div class="line">  my ( $self, $other_alignments ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $other_alignments ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a9a6e4ac2019911b27e719d1c794de0c8">other_alignments</a>}     = $other_alignments;</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aae89a27e1bf3e2b99f5ed03481d1c5dd">has_other_alignments</a>} = undef;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a9a6e4ac2019911b27e719d1c794de0c8">other_alignments</a>} &amp;&amp;</div>
<div class="line">          defined $self-&gt;adaptor() )</div>
<div class="line">  {</div>
<div class="line">    $self-&gt;adaptor()-&gt;_fetch_other_alignments($self);</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a9a6e4ac2019911b27e719d1c794de0c8">other_alignments</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a1cc2eddd90cfea6cdba23b8498192b2f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::GenomeInfo::publications </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) arrayref of pubmed IDs to set
  Description: Gets/sets PubMed IDs for publications associated with the genome
  Returntype : Arrayref of PubMed IDs
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-publications" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-publications-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-publications-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-publications-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a1cc2eddd90cfea6cdba23b8498192b2f">publications</a> {</div>
<div class="line">  my ( $self, $publications ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $publications ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a1cc2eddd90cfea6cdba23b8498192b2f">publications</a>} = $publications;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a1cc2eddd90cfea6cdba23b8498192b2f">publications</a>} &amp;&amp; defined $self-&gt;adaptor() ) {</div>
<div class="line">    $self-&gt;adaptor()-&gt;_fetch_publications($self);</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a1cc2eddd90cfea6cdba23b8498192b2f">publications</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a537e48c3d95f8a42927fcff0d20faaa9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::GenomeInfo::sequences </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) arrayref of sequences to set
  Description: Gets/sets array of hashrefs describing sequences from the assembly. Elements are hashrefs with name and acc as keys
  Returntype : Arrayref
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-sequences" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-sequences-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-sequences-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-sequences-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a537e48c3d95f8a42927fcff0d20faaa9">sequences</a> {</div>
<div class="line">  my ( $self, $sequences ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $sequences ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a537e48c3d95f8a42927fcff0d20faaa9">sequences</a>} = $sequences;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a537e48c3d95f8a42927fcff0d20faaa9">sequences</a>} &amp;&amp; defined $self-&gt;adaptor() ) {</div>
<div class="line">    $self-&gt;adaptor()-&gt;_fetch_sequences($self);</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a537e48c3d95f8a42927fcff0d20faaa9">sequences</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aaadeee2f4c9733f317389fd57d9a08be"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::serotype </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) serotype to set
  Description: Gets/sets serotype
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-serotype" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-serotype-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-serotype-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-serotype-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aaadeee2f4c9733f317389fd57d9a08be">serotype</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aaadeee2f4c9733f317389fd57d9a08be">serotype</a>} = $arg <span class="keywordflow">if</span> ( defined $arg );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aaadeee2f4c9733f317389fd57d9a08be">serotype</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac22dff02f6bb3bd3cce06c06c1afdc3d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::GenomeInfo::species </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-species" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-species-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-species-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-species-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac22dff02f6bb3bd3cce06c06c1afdc3d">species</a> {</div>
<div class="line">  my ( $self, $species ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac22dff02f6bb3bd3cce06c06c1afdc3d">species</a>} = $species <span class="keywordflow">if</span> ( defined $species );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#ac22dff02f6bb3bd3cce06c06c1afdc3d">species</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a4e4e9b068c87106f44d9bd6096b6edec"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::species_id </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) species_id to set
  Description: Gets/sets species_id of genome within core database
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-species_id" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-species_id-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-species_id-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-species_id-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4e4e9b068c87106f44d9bd6096b6edec">species_id</a> {</div>
<div class="line">  my ( $self, $species_id ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4e4e9b068c87106f44d9bd6096b6edec">species_id</a>} = $species_id <span class="keywordflow">if</span> ( defined $species_id );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a4e4e9b068c87106f44d9bd6096b6edec">species_id</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="af078b69ba449d043cceae3bf3a69272e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::strain </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) strain to set
  Description: Gets/sets strain of genome
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-strain" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-strain-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-strain-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-strain-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af078b69ba449d043cceae3bf3a69272e">strain</a> {</div>
<div class="line">  my ( $self, $arg ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af078b69ba449d043cceae3bf3a69272e">strain</a>} = $arg <span class="keywordflow">if</span> ( defined $arg );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#af078b69ba449d043cceae3bf3a69272e">strain</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a071717879973be03c76374af3f1e582a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public String Bio::EnsEMBL::Utils::MetaData::GenomeInfo::taxonomy_id </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) taxonomy_id to set
  Description: Gets/sets NCBI taxonomy ID
  Returntype : string
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-taxonomy_id" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-taxonomy_id-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-taxonomy_id-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-taxonomy_id-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a071717879973be03c76374af3f1e582a">taxonomy_id</a> {</div>
<div class="line">  my ( $self, $taxonomy_id ) = @_;</div>
<div class="line">  $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a071717879973be03c76374af3f1e582a">taxonomy_id</a>} = $taxonomy_id <span class="keywordflow">if</span> ( defined $taxonomy_id );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a071717879973be03c76374af3f1e582a">taxonomy_id</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aa88bd8a7e5fb478fbd77d1d1db3ce841"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hashref Bio::EnsEMBL::Utils::MetaData::GenomeInfo::to_hash </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Description: Render genome as plain hash suitable for export as JSON/XML
  Argument   : (optional) if set to 1, force expansion of children
  Returntype : Hashref
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-to_hash" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-to_hash-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-to_hash-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-to_hash-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aa88bd8a7e5fb478fbd77d1d1db3ce841">to_hash</a> {</div>
<div class="line">  my ( $in, $keen ) = @_;</div>
<div class="line">  my $out;</div>
<div class="line">  my $type = ref $in;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $keen &amp;&amp;</div>
<div class="line">       $keen == 1 &amp;&amp;</div>
<div class="line">       $type eq <span class="stringliteral">&#39;Bio::EnsEMBL::Utils::MetaData::GenomeInfo&#39;</span> )</div>
<div class="line">  {</div>
<div class="line">    $in-&gt;_preload();</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">if</span> ( $type eq <span class="stringliteral">&#39;ARRAY&#39;</span> ) {</div>
<div class="line">    $out = [];</div>
<div class="line">    <span class="keywordflow">for</span> my $item ( @{$in} ) {</div>
<div class="line">      push @{$out}, <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aa88bd8a7e5fb478fbd77d1d1db3ce841">to_hash</a>( $item, $keen );</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  elsif ( $type eq <span class="stringliteral">&#39;HASH&#39;</span> ||</div>
<div class="line">          $type eq <span class="stringliteral">&#39;Bio::EnsEMBL::Utils::MetaData::GenomeInfo&#39;</span> )</div>
<div class="line">  {</div>
<div class="line">    $out = {};</div>
<div class="line">    <span class="keywordflow">while</span> ( my ( $key, $val ) = each %$in ) {</div>
<div class="line">      <span class="keywordflow">if</span> ( $key ne <span class="stringliteral">&#39;dbID&#39;</span> &amp;&amp; $key ne <span class="stringliteral">&#39;adaptor&#39;</span> &amp;&amp; $key ne <span class="stringliteral">&#39;logger&#39;</span> ) {</div>
<div class="line"><span class="preprocessor"># deal with keys starting with numbers, which are not valid element names in XML</span></div>
<div class="line"><span class="preprocessor"></span>        <span class="keywordflow">if</span> ( $key =~ m/^[0-9].*/ ) {</div>
<div class="line">          $key = <span class="charliteral">&#39;_&#39;</span> . $key;</div>
<div class="line">        }</div>
<div class="line">        $out-&gt;{$key} = <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aa88bd8a7e5fb478fbd77d1d1db3ce841">to_hash</a>( $val, $keen );</div>
<div class="line">      }</div>
<div class="line"></div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  elsif ( $type eq <span class="stringliteral">&#39;Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo&#39;</span> )</div>
<div class="line">  {</div>
<div class="line">    $out = $in-&gt;to_hash();</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">else</span> {</div>
<div class="line">    $out = $in;</div>
<div class="line">  }</div>
<div class="line"></div>
<div class="line">  <span class="keywordflow">return</span> $out;</div>
<div class="line">} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#aa88bd8a7e5fb478fbd77d1d1db3ce841">to_hash</a></div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a03cc3fba9fa7a7cd24ad172c45ac554b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::GenomeInfo::variations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) variations to set
  Description: Gets/sets variations associated with genomes as hashref 
               (variations,structural variations,genotypes,phenotypes), 
               further broken down into counts by type/source
  Returntype : Arrayref
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-variations" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-variations-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-variations-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-variations-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a03cc3fba9fa7a7cd24ad172c45ac554b">variations</a> {</div>
<div class="line">  my ( $self, $variations ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $variations ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a03cc3fba9fa7a7cd24ad172c45ac554b">variations</a>}     = $variations;</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a346b3ac487b2732b3f3a76bf7a657618">has_variations</a>} = undef;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a03cc3fba9fa7a7cd24ad172c45ac554b">variations</a>} &amp;&amp; defined $self-&gt;adaptor() ) {</div>
<div class="line">    $self-&gt;adaptor()-&gt;_fetch_variations($self);</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html#a03cc3fba9fa7a7cd24ad172c45ac554b">variations</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>modules/Bio/EnsEMBL/Utils/MetaData/GenomeInfo.pm</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.3.1
</small></address>
</body>
</html>
