<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.3.1"/>
<title>Ensembl Genomes API: Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">Ensembl Genomes API
   </div>
   <div id="projectbrief">Auxillary Perl API for Ensembl Genomes</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.3.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>Bio</b></li><li class="navelem"><b>EnsEMBL</b></li><li class="navelem"><b>Utils</b></li><li class="navelem"><b>MetaData</b></li><li class="navelem"><b>DBSQL</b></li><li class="navelem"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">GenomeInfoAdaptor</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ada57d4b94bf938786da91eb2b97a754b"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ada57d4b94bf938786da91eb2b97a754b">new</a> ()</td></tr>
<tr class="separator:ada57d4b94bf938786da91eb2b97a754b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab92da85659d7ffd46da7bad148d448b4"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab92da85659d7ffd46da7bad148d448b4">build_adaptor</a> ()</td></tr>
<tr class="separator:ab92da85659d7ffd46da7bad148d448b4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a21e83f339f3d250c860f768d68e7e04f"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a21e83f339f3d250c860f768d68e7e04f">store</a> ()</td></tr>
<tr class="separator:a21e83f339f3d250c860f768d68e7e04f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a36a75246ac8a7ac25a2c8ac74ae0192e"><td class="memItemLeft" align="right" valign="top">public void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a36a75246ac8a7ac25a2c8ac74ae0192e">update</a> ()</td></tr>
<tr class="separator:a36a75246ac8a7ac25a2c8ac74ae0192e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2b9a46d26f3bb8b3395888d2a5dedf23"><td class="memItemLeft" align="right" valign="top">public void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a2b9a46d26f3bb8b3395888d2a5dedf23">update_compara</a> ()</td></tr>
<tr class="separator:a2b9a46d26f3bb8b3395888d2a5dedf23"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad8fbb82f254c61dcc029449f5f048ba"><td class="memItemLeft" align="right" valign="top">public void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aad8fbb82f254c61dcc029449f5f048ba">store_compara</a> ()</td></tr>
<tr class="separator:aad8fbb82f254c61dcc029449f5f048ba"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb913c752b8a9bf2d4c29741634e5e6c"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#adb913c752b8a9bf2d4c29741634e5e6c">_store_aliases</a> ()</td></tr>
<tr class="separator:adb913c752b8a9bf2d4c29741634e5e6c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a49ed7b29d3e75215103dceaade9ad455"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a49ed7b29d3e75215103dceaade9ad455">_store_sequences</a> ()</td></tr>
<tr class="separator:a49ed7b29d3e75215103dceaade9ad455"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a41218a150218370f0c1cdd75ee8812a6"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a41218a150218370f0c1cdd75ee8812a6">_store_features</a> ()</td></tr>
<tr class="separator:a41218a150218370f0c1cdd75ee8812a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5bbc6aeefb9a713e23c7402019086034"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a5bbc6aeefb9a713e23c7402019086034">_store_annotations</a> ()</td></tr>
<tr class="separator:a5bbc6aeefb9a713e23c7402019086034"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a00fc0551c262bce91e305ca859325738"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a00fc0551c262bce91e305ca859325738">_store_variations</a> ()</td></tr>
<tr class="separator:a00fc0551c262bce91e305ca859325738"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad80eff92443d9527737e8883b149cd7b"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ad80eff92443d9527737e8883b149cd7b">_store_alignments</a> ()</td></tr>
<tr class="separator:ad80eff92443d9527737e8883b149cd7b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af138803c65700ed94423cf35c63109cd"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af138803c65700ed94423cf35c63109cd">taxonomy_adaptor</a> ()</td></tr>
<tr class="separator:af138803c65700ed94423cf35c63109cd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0c4bba091ca39f03f2ebdbc5313118f9"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a0c4bba091ca39f03f2ebdbc5313118f9">list_divisions</a> ()</td></tr>
<tr class="separator:a0c4bba091ca39f03f2ebdbc5313118f9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a98fec113b5f7a90ebfc7473692967325"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a98fec113b5f7a90ebfc7473692967325">fetch_all</a> ()</td></tr>
<tr class="separator:a98fec113b5f7a90ebfc7473692967325"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac2d36ac6d8bd591f0737224575231d27"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac2d36ac6d8bd591f0737224575231d27">fetch_by_dbID</a> ()</td></tr>
<tr class="separator:ac2d36ac6d8bd591f0737224575231d27"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad8aaa0c16c88e0223e80907e03dbc4de"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ad8aaa0c16c88e0223e80907e03dbc4de">fetch_by_dbIDs</a> ()</td></tr>
<tr class="separator:ad8aaa0c16c88e0223e80907e03dbc4de"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab3ebe74e0af6996ddd371bd32d9bbe2a"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab3ebe74e0af6996ddd371bd32d9bbe2a">fetch_all_by_sequence_accession</a> ()</td></tr>
<tr class="separator:ab3ebe74e0af6996ddd371bd32d9bbe2a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac32ebacac51ec0d2b59db0b573b61c2a"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac32ebacac51ec0d2b59db0b573b61c2a">fetch_all_by_sequence_accession_unversioned</a> ()</td></tr>
<tr class="separator:ac32ebacac51ec0d2b59db0b573b61c2a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa57e37a0374bd7a0de661d4c9f5265a4"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa57e37a0374bd7a0de661d4c9f5265a4">fetch_all_by_sequence_accession_versioned</a> ()</td></tr>
<tr class="separator:aa57e37a0374bd7a0de661d4c9f5265a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a96453090c2be26ed6664d24f3b09b53d"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a96453090c2be26ed6664d24f3b09b53d">fetch_by_assembly_id</a> ()</td></tr>
<tr class="separator:a96453090c2be26ed6664d24f3b09b53d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af3e0556cf0741089d1c3be618ae01e1c"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af3e0556cf0741089d1c3be618ae01e1c">fetch_by_assembly_id_versioned</a> ()</td></tr>
<tr class="separator:af3e0556cf0741089d1c3be618ae01e1c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae4e1603ff44789591809234c959bc335"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ae4e1603ff44789591809234c959bc335">fetch_by_assembly_id_unversioned</a> ()</td></tr>
<tr class="separator:ae4e1603ff44789591809234c959bc335"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afcb2ec45290da01dea42ee09e9148e7b"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#afcb2ec45290da01dea42ee09e9148e7b">fetch_all_by_taxonomy_id</a> ()</td></tr>
<tr class="separator:afcb2ec45290da01dea42ee09e9148e7b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aff91c234f1f9358536087f5a8ba5b251"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aff91c234f1f9358536087f5a8ba5b251">fetch_all_by_taxonomy_ids</a> ()</td></tr>
<tr class="separator:aff91c234f1f9358536087f5a8ba5b251"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac85701101dbecf4b1f5c58574dcb2f2c"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac85701101dbecf4b1f5c58574dcb2f2c">fetch_all_by_taxonomy_branch</a> ()</td></tr>
<tr class="separator:ac85701101dbecf4b1f5c58574dcb2f2c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7d9aa9d28b19da134c14ab77a85f6c23"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a7d9aa9d28b19da134c14ab77a85f6c23">fetch_all_by_division</a> ()</td></tr>
<tr class="separator:a7d9aa9d28b19da134c14ab77a85f6c23"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab05b21c4e09665860c3fb5ea60579cbd"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab05b21c4e09665860c3fb5ea60579cbd">fetch_by_species</a> ()</td></tr>
<tr class="separator:ab05b21c4e09665860c3fb5ea60579cbd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae544a9ec503f127cd25a3b9bfdf0a6ad"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ae544a9ec503f127cd25a3b9bfdf0a6ad">fetch_by_name</a> ()</td></tr>
<tr class="separator:ae544a9ec503f127cd25a3b9bfdf0a6ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1e18e1443071bedce172e383ce7ab033"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a1e18e1443071bedce172e383ce7ab033">fetch_by_any_name</a> ()</td></tr>
<tr class="separator:a1e18e1443071bedce172e383ce7ab033"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8ff7541994238568529e5dace97b299"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac8ff7541994238568529e5dace97b299">fetch_all_by_dbname</a> ()</td></tr>
<tr class="separator:ac8ff7541994238568529e5dace97b299"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab05c482f397d29a8dc3993a8c7c3113d"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab05c482f397d29a8dc3993a8c7c3113d">fetch_all_by_name_pattern</a> ()</td></tr>
<tr class="separator:ab05c482f397d29a8dc3993a8c7c3113d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac2dffbf3e25b19e11a9f0432ba42a655"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac2dffbf3e25b19e11a9f0432ba42a655">fetch_by_alias</a> ()</td></tr>
<tr class="separator:ac2dffbf3e25b19e11a9f0432ba42a655"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a92ec339a777775b5a771791f47189328"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a92ec339a777775b5a771791f47189328">fetch_all_with_variation</a> ()</td></tr>
<tr class="separator:a92ec339a777775b5a771791f47189328"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac55295a8f575021917f6e2ac61f4d18c"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac55295a8f575021917f6e2ac61f4d18c">fetch_all_with_peptide_compara</a> ()</td></tr>
<tr class="separator:ac55295a8f575021917f6e2ac61f4d18c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab8567f2e206b6c653ed1163283c6974e"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab8567f2e206b6c653ed1163283c6974e">fetch_all_with_pan_compara</a> ()</td></tr>
<tr class="separator:ab8567f2e206b6c653ed1163283c6974e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa69042b2065f079af54aa363f916f3e2"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa69042b2065f079af54aa363f916f3e2">fetch_all_with_genome_alignments</a> ()</td></tr>
<tr class="separator:aa69042b2065f079af54aa363f916f3e2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0fed890cc2d6b3a5a1452b6873c19d0c"><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a0fed890cc2d6b3a5a1452b6873c19d0c">fetch_all_with_compara</a> ()</td></tr>
<tr class="separator:a0fed890cc2d6b3a5a1452b6873c19d0c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0f9ead633abe1821d805c70def8e0f9b"><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a0f9ead633abe1821d805c70def8e0f9b">fetch_all_with_other_alignments</a> ()</td></tr>
<tr class="separator:a0f9ead633abe1821d805c70def8e0f9b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa584d1c91e4b551a6cf2e6f8a02d18f3"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa584d1c91e4b551a6cf2e6f8a02d18f3">_fetch_variations</a> ()</td></tr>
<tr class="separator:aa584d1c91e4b551a6cf2e6f8a02d18f3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a788d3fccde1fca75116e4a08576589f2"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a788d3fccde1fca75116e4a08576589f2">_fetch_other_alignments</a> ()</td></tr>
<tr class="separator:a788d3fccde1fca75116e4a08576589f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42181464c90aacc8cc05675f19581b94"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a42181464c90aacc8cc05675f19581b94">_fetch_annotations</a> ()</td></tr>
<tr class="separator:a42181464c90aacc8cc05675f19581b94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adc118d7d9d61496c95ee155f4ed9df5f"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#adc118d7d9d61496c95ee155f4ed9df5f">_fetch_features</a> ()</td></tr>
<tr class="separator:adc118d7d9d61496c95ee155f4ed9df5f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac45ecaf64a7ae889a88cf6bb493ec739"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac45ecaf64a7ae889a88cf6bb493ec739">_fetch_publications</a> ()</td></tr>
<tr class="separator:ac45ecaf64a7ae889a88cf6bb493ec739"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad927a3bd07558adf0d5e9166fd1d4648"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ad927a3bd07558adf0d5e9166fd1d4648">_fetch_aliases</a> ()</td></tr>
<tr class="separator:ad927a3bd07558adf0d5e9166fd1d4648"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1af91719c30050bf0075ec6d9e3b73bd"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a1af91719c30050bf0075ec6d9e3b73bd">_fetch_sequences</a> ()</td></tr>
<tr class="separator:a1af91719c30050bf0075ec6d9e3b73bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab7423e335837c98640573945025a13c8"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab7423e335837c98640573945025a13c8">_fetch_comparas</a> ()</td></tr>
<tr class="separator:ab7423e335837c98640573945025a13c8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af2d425b3fd6e7d0188ead526bad562bf"><td class="memItemLeft" align="right" valign="top">public Array&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af2d425b3fd6e7d0188ead526bad562bf">fetch_all_comparas</a> ()</td></tr>
<tr class="separator:af2d425b3fd6e7d0188ead526bad562bf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeab1847847a9eece525cd7b8c88a4ddb"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aeab1847847a9eece525cd7b8c88a4ddb">fetch_compara_by_dbID</a> ()</td></tr>
<tr class="separator:aeab1847847a9eece525cd7b8c88a4ddb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7407979a544da19c04c26f0e3dc4b215"><td class="memItemLeft" align="right" valign="top">public Array&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a7407979a544da19c04c26f0e3dc4b215">fetch_all_compara_by_division</a> ()</td></tr>
<tr class="separator:a7407979a544da19c04c26f0e3dc4b215"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac4eaa65a71d64e2e90b5a1469e09b8d4"><td class="memItemLeft" align="right" valign="top">public Array&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac4eaa65a71d64e2e90b5a1469e09b8d4">fetch_all_compara_by_method</a> ()</td></tr>
<tr class="separator:ac4eaa65a71d64e2e90b5a1469e09b8d4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9c54cedfdafaf641a866554bd7e1f70f"><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a9c54cedfdafaf641a866554bd7e1f70f">fetch_compara_by_dbname_method_set</a> ()</td></tr>
<tr class="separator:a9c54cedfdafaf641a866554bd7e1f70f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaac7e9dfb49a4c536ad83beba79baf66"><td class="memItemLeft" align="right" valign="top">protected Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aaac7e9dfb49a4c536ad83beba79baf66">_fetch_compara_with_args</a> ()</td></tr>
<tr class="separator:aaac7e9dfb49a4c536ad83beba79baf66"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa2a0975d5a4ccdb98c49f66cf1d02542"><td class="memItemLeft" align="right" valign="top">protected Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa2a0975d5a4ccdb98c49f66cf1d02542">_fetch_compara_generic</a> ()</td></tr>
<tr class="separator:aa2a0975d5a4ccdb98c49f66cf1d02542"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8cb5669c8ee524ffc704477321c7b961"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a8cb5669c8ee524ffc704477321c7b961">_fetch_children</a> ()</td></tr>
<tr class="separator:a8cb5669c8ee524ffc704477321c7b961"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a44a14301f9e35e3cf9829c8f42babf78"><td class="memItemLeft" align="right" valign="top">protected Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a44a14301f9e35e3cf9829c8f42babf78">_fetch_generic_with_args</a> ()</td></tr>
<tr class="separator:a44a14301f9e35e3cf9829c8f42babf78"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3a2b1e2023eff92e02ece082bcb2fbc3"><td class="memItemLeft" align="right" valign="top">protected Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a3a2b1e2023eff92e02ece082bcb2fbc3">_fetch_generic</a> ()</td></tr>
<tr class="separator:a3a2b1e2023eff92e02ece082bcb2fbc3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3ffe9111f20a57dc6ea13149e0b18b2f"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a3ffe9111f20a57dc6ea13149e0b18b2f">_cache</a> ()</td></tr>
<tr class="separator:a3ffe9111f20a57dc6ea13149e0b18b2f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aff6d39c4d0db9b777495d95e291940e2"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aff6d39c4d0db9b777495d95e291940e2">_clear_cache</a> ()</td></tr>
<tr class="separator:aff6d39c4d0db9b777495d95e291940e2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1439390262ce0abc462135515ed7922d"><td class="memItemLeft" align="right" valign="top">protected Object&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a1439390262ce0abc462135515ed7922d">_get_cached_obj</a> ()</td></tr>
<tr class="separator:a1439390262ce0abc462135515ed7922d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae3fc01bba18f5e5ab7a3d34f8849c1c5"><td class="memItemLeft" align="right" valign="top">protected void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ae3fc01bba18f5e5ab7a3d34f8849c1c5">_store_cached_obj</a> ()</td></tr>
<tr class="separator:ae3fc01bba18f5e5ab7a3d34f8849c1c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa1b982df106fad0f575052e5023a45c2"><td class="memItemLeft" align="right" valign="top">protected Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a> ()</td></tr>
<tr class="separator:aa1b982df106fad0f575052e5023a45c2"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h1><a class="anchor" id="Synopsis"></a>
Synopsis</h1>
<div class="fragment"><div class="line">my $gdba = <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab92da85659d7ffd46da7bad148d448b4">build_adaptor</a>();</div>
<div class="line">my $md = $gdba-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab05b21c4e09665860c3fb5ea60579cbd">fetch_by_species</a>(<span class="stringliteral">&quot;arabidopsis_thaliana&quot;</span>);</div>
</div><!-- fragment --> <h1><a class="anchor" id="Description"></a>
Description</h1>
 <pre>
Adaptor for storing and retrieving GenomeInfo objects from MySQL genome_info database

To start working with an adaptor:

# getting an adaptor
## adaptor for latest public EG release
my $gdba = Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor->build_adaptor();
## adaptor for specified public EG release
my $gdba = Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor->build_adaptor(21);
## manually specify a given database
my $dbc = Bio::EnsEMBL::DBSQL::DBConnection->new(
-USER=>'anonymous',
-PORT=>4157,
-HOST=>'mysql-eg-publicsql.ebi.ac.uk',
-DBNAME=>'genome_info_21');
my $gdba = Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor->new(-DBC=>$dbc);

To find genomes, use the fetch methods e.g.

# find a genome by name
my $genome = $gdba->fetch_by_species('arabidopsis_thaliana');

# find and iterate over all genomes
for my $genome (\@{$gdba->fetch_all()}) {
    print $genome->name()."\n";
}

# find and iterate over all genomes from plants
for my $genome (\@{$gdba->fetch_all_by_division('EnsemblPlants')}) {
    print $genome->name()."\n";
}

# find and iterate over all genomes with variation
for my $genome (\@{$gdba->fetch_all_with_variation()}) {
    print $genome->name()."\n";
}

# find all comparas for the division of interest
my $comparas = $gdba->fetch_all_compara_by_division('EnsemblPlants');

# find the peptide compara
my ($compara) = grep {$_->is_peptide_compara()} \@$comparas;
print $compara->division()." ".$compara->method()."(".$compara->dbname().")\n";

# print out all the genomes in this compara
for my $genome (\@{$compara->genomes()}) {
    print $genome->name()."\n";
}

</pre> </div><h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a3ffe9111f20a57dc6ea13149e0b18b2f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_cache </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : type of object for cache
  Description: Return internal cache for given type
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_cache" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_cache-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_cache-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_cache-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a3ffe9111f20a57dc6ea13149e0b18b2f">_cache</a> {</div>
<div class="line">  my ( $self, $type ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( !defined $self-&gt;{cache} || !defined $self-&gt;{cache}{$type} ) {</div>
<div class="line">    $self-&gt;{cache}{$type} = {};</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{cache}{$type};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aff6d39c4d0db9b777495d95e291940e2"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_clear_cache </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : (optional) type of object to clear
  Description: Clear internal cache (optionally just one type)
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_clear_cache" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_clear_cache-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_clear_cache-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_clear_cache-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aff6d39c4d0db9b777495d95e291940e2">_clear_cache</a> {</div>
<div class="line">  my ( $self, $type ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $type ) {</div>
<div class="line">    $self-&gt;{cache}{$type} = {};</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">else</span> {</div>
<div class="line">    $self-&gt;{cache} = {};</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ad927a3bd07558adf0d5e9166fd1d4648"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_aliases </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> 
  Description: Add aliases to supplied object
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_aliases" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_aliases-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_aliases-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_aliases-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ad927a3bd07558adf0d5e9166fd1d4648">_fetch_aliases</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  croak</div>
<div class="line"><span class="stringliteral">&quot;Cannot fetch aliases for a GenomeInfo object that has not been stored&quot;</span></div>
<div class="line">    <span class="keywordflow">if</span> !defined $genome-&gt;dbID();</div>
<div class="line">  my $aliases =</div>
<div class="line">    $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_simple(</div>
<div class="line">             -SQL =&gt; <span class="stringliteral">&#39;select alias from genome_alias where genome_id=?&#39;</span>,</div>
<div class="line">             -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line">  $genome-&gt;aliases($aliases);</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a42181464c90aacc8cc05675f19581b94"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_annotations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> 
  Description: Add annotations to supplied object
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_annotations" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_annotations-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_annotations-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_annotations-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a42181464c90aacc8cc05675f19581b94">_fetch_annotations</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  croak</div>
<div class="line"><span class="stringliteral">&quot;Cannot fetch annotations for a GenomeInfo object that has not been stored&quot;</span></div>
<div class="line">    <span class="keywordflow">if</span> !defined $genome-&gt;dbID();</div>
<div class="line">  my $annotations = {};</div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_no_return(</div>
<div class="line">    -SQL =&gt;</div>
<div class="line">      <span class="stringliteral">&#39;select type,count from genome_annotation where genome_id=?&#39;</span>,</div>
<div class="line">    -CALLBACK =&gt; sub {</div>
<div class="line">      my @row = @{ shift @_ };</div>
<div class="line">      $annotations-&gt;{ $row[0] } = $row[1];</div>
<div class="line">      <span class="keywordflow">return</span>;</div>
<div class="line">    },</div>
<div class="line">    -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line">  $genome-&gt;annotations($annotations);</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a8cb5669c8ee524ffc704477321c7b961"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_children </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : Arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Description: Fetch all children of specified genome info object
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_children" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_children-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_children-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_children-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a8cb5669c8ee524ffc704477321c7b961">_fetch_children</a> {</div>
<div class="line">  my ( $self, $md ) = @_;</div>
<div class="line">  $self-&gt;_fetch_variations($md);</div>
<div class="line">  $self-&gt;_fetch_sequences($md);</div>
<div class="line">  $self-&gt;_fetch_aliases($md);</div>
<div class="line">  $self-&gt;_fetch_publications($md);</div>
<div class="line">  $self-&gt;_fetch_annotations($md);</div>
<div class="line">  $self-&gt;_fetch_other_alignments($md);</div>
<div class="line">  $self-&gt;_fetch_comparas($md);</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aa2a0975d5a4ccdb98c49f66cf1d02542"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_compara_generic </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : SQL
  Arg        : array ref of bind params
  Description: Fetch compara analyses using supplied SQL
  Returntype : arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_compara_generic" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_compara_generic-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_compara_generic-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_compara_generic-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa2a0975d5a4ccdb98c49f66cf1d02542">_fetch_compara_generic</a> {</div>
<div class="line">  my ( $self, $sql, $params ) = @_;</div>
<div class="line"><span class="preprocessor">  # check to see if we&#39;ve cached this already</span></div>
<div class="line"><span class="preprocessor"></span></div>
<div class="line"><span class="preprocessor">  # build the main object first</span></div>
<div class="line"><span class="preprocessor"></span>  my $comparas = $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute(</div>
<div class="line">    -SQL      =&gt; $sql,</div>
<div class="line">    -PARAMS   =&gt; $params,</div>
<div class="line">    -CALLBACK =&gt; sub {</div>
<div class="line">      my @row = @{ shift @_ };</div>
<div class="line">      my $id  = $row[0];</div>
<div class="line">      my $c = $self-&gt;_get_cached_obj(</div>
<div class="line">                     <span class="stringliteral">&#39;Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo&#39;</span>,</div>
<div class="line">                     $id );</div>
<div class="line">      <span class="keywordflow">if</span> ( !defined $c ) {</div>
<div class="line">        $c = <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html#a82107ea5bf954e94898365301da99e4d">new</a>();</div>
<div class="line">        $c-&gt;dbID($id);</div>
<div class="line">        $c-&gt;division( $row[1] );</div>
<div class="line">        $c-&gt;method( $row[2] );</div>
<div class="line">        $c-&gt;set_name( $row[3] );</div>
<div class="line">        $c-&gt;dbname( $row[4] );</div>
<div class="line"><span class="preprocessor">        # cache the completed compara object</span></div>
<div class="line"><span class="preprocessor"></span>        $c-&gt;adaptor($self);</div>
<div class="line">        $self-&gt;_store_cached_obj(</div>
<div class="line">                     <span class="stringliteral">&#39;Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo&#39;</span>,</div>
<div class="line">                     $c );</div>
<div class="line">      }</div>
<div class="line">      <span class="keywordflow">return</span> $c;</div>
<div class="line">    } );</div>
<div class="line">  <span class="keywordflow">for</span> my $compara (@$comparas) {</div>
<div class="line"><span class="preprocessor"># add genomes on one by one (don&#39;t nest the fetch here as could run of connections)</span></div>
<div class="line"><span class="preprocessor"></span>    <span class="keywordflow">if</span> ( !defined $compara-&gt;{genomes} ) {</div>
<div class="line">      my $genomes = [];</div>
<div class="line">      <span class="keywordflow">for</span> my $genome_id (</div>
<div class="line">        @{$self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_simple(</div>
<div class="line">            -SQL =&gt;</div>
<div class="line">q/select distinct(genome_id) from genome_compara_analysis where compara_analysis_id=?/,</div>
<div class="line">            -PARAMS =&gt; [ $compara-&gt;dbID() ] );</div>
<div class="line">        } )</div>
<div class="line">      {</div>
<div class="line">        push @$genomes, $self-&gt;fetch_by_dbID($genome_id);</div>
<div class="line">      }</div>
<div class="line">      $compara-&gt;genomes($genomes);</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $comparas;</div>
<div class="line">} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa2a0975d5a4ccdb98c49f66cf1d02542">_fetch_compara_generic</a></div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aaac7e9dfb49a4c536ad83beba79baf66"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_compara_with_args </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : hashref of arguments by column
  Description: Fetch compara analyses by column value pairs
  Returntype : arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_compara_with_args" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_compara_with_args-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_compara_with_args-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_compara_with_args-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aaac7e9dfb49a4c536ad83beba79baf66">_fetch_compara_with_args</a> {</div>
<div class="line">  my ( $self, $args ) = @_;</div>
<div class="line">  my $sql    = $base_compara_fetch_sql;</div>
<div class="line">  my $params = [ values %$args ];</div>
<div class="line">  my $clause = join( <span class="stringliteral">&#39; AND &#39;</span>, map { $_ . <span class="stringliteral">&#39;=?&#39;</span> } keys %$args );</div>
<div class="line">  <span class="keywordflow">if</span> ( $clause ne <span class="stringliteral">&#39;&#39;</span> ) {</div>
<div class="line">    $sql .= <span class="stringliteral">&#39; where &#39;</span> . $clause;</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_compara_generic( $sql, $params );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ab7423e335837c98640573945025a13c8"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_comparas </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> 
  Description: Add compara info to supplied object
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_comparas" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_comparas-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_comparas-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_comparas-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab7423e335837c98640573945025a13c8">_fetch_comparas</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  my $comparas = [];</div>
<div class="line">  <span class="keywordflow">for</span> my $id (</div>
<div class="line">    @{$self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_simple(</div>
<div class="line">        -SQL =&gt;</div>
<div class="line">          q/select distinct compara_analysis_id from compara_analysis </div>
<div class="line">  join genome_compara_analysis <span class="keyword">using</span> (compara_analysis_id)</div>
<div class="line">  where genome_id=? and method in (<span class="stringliteral">&#39;BLASTZ_NET&#39;</span>,<span class="stringliteral">&#39;LASTZ_NET&#39;</span>,<span class="stringliteral">&#39;TRANSLATED_BLAT_NET&#39;</span>, <span class="stringliteral">&#39;PROTEIN_TREES&#39;</span>)/,</div>
<div class="line">        -PARAMS =&gt; [ $genome-&gt;dbID() ] ) } )</div>
<div class="line">  {</div>
<div class="line">    push @$comparas, $self-&gt;fetch_compara_by_dbID($id);</div>
<div class="line">  }</div>
<div class="line">  $genome-&gt;compara($comparas);</div>
<div class="line">  $genome-&gt;has_pan_compara();</div>
<div class="line">  $genome-&gt;has_genome_alignments();</div>
<div class="line">  $genome-&gt;has_peptide_compara();</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="adc118d7d9d61496c95ee155f4ed9df5f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_features </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> 
  Description: Add features to supplied object
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_features" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_features-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_features-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_features-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#adc118d7d9d61496c95ee155f4ed9df5f">_fetch_features</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  croak</div>
<div class="line"><span class="stringliteral">&quot;Cannot fetch features  for a GenomeInfo object that has not been stored&quot;</span></div>
<div class="line">    <span class="keywordflow">if</span> !defined $genome-&gt;dbID();</div>
<div class="line">  my $features = {};</div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_no_return(</div>
<div class="line">    -SQL =&gt;</div>
<div class="line"><span class="stringliteral">&#39;select type,analysis,count from genome_feature where genome_id=?&#39;</span>,</div>
<div class="line">    -CALLBACK =&gt; sub {</div>
<div class="line">      my @row = @{ shift @_ };</div>
<div class="line">      $features-&gt;{ $row[0] }-&gt;{ $row[1] } = $row[2];</div>
<div class="line">      <span class="keywordflow">return</span>;</div>
<div class="line">    },</div>
<div class="line">    -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line">  $genome-&gt;features($features);</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a3a2b1e2023eff92e02ece082bcb2fbc3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_generic </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : SQL to use to fetch object
  Arg        : arrayref of bind parameters
  Arg        : (optional) if set to 1, all children will be fetched
  Description: Instantiate a <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">GenomeInfo</a> from the database using the specified SQL
  Returntype : Arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_generic" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_generic-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_generic-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_generic-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a3a2b1e2023eff92e02ece082bcb2fbc3">_fetch_generic</a> {</div>
<div class="line">  my ( $self, $sql, $params, $keen ) = @_;</div>
<div class="line">  my $mds = $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute(</div>
<div class="line">    -SQL          =&gt; $sql,</div>
<div class="line">    -USE_HASHREFS =&gt; 1,</div>
<div class="line">    -CALLBACK     =&gt; sub {</div>
<div class="line">      my $row = shift @_;</div>
<div class="line">      my $md =</div>
<div class="line">        $self-&gt;_get_cached_obj(</div>
<div class="line">                            <span class="stringliteral">&#39;Bio::EnsEMBL::Utils::MetaData::GenomeInfo&#39;</span>,</div>
<div class="line">                            $row-&gt;{dbID} );</div>
<div class="line">      <span class="keywordflow">if</span> ( !defined $md ) {</div>
<div class="line">        $md = bless $row, <span class="stringliteral">&#39;Bio::EnsEMBL::Utils::MetaData::GenomeInfo&#39;</span>;</div>
<div class="line">        $md-&gt;adaptor($self);</div>
<div class="line">        $self-&gt;_store_cached_obj(</div>
<div class="line">                            <span class="stringliteral">&#39;Bio::EnsEMBL::Utils::MetaData::GenomeInfo&#39;</span>,</div>
<div class="line">                            $md );</div>
<div class="line">      }</div>
<div class="line">      <span class="keywordflow">return</span> $md;</div>
<div class="line">    },</div>
<div class="line">    -PARAMS =&gt; $params );</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $keen &amp;&amp; $keen == 1 ) {</div>
<div class="line">    <span class="keywordflow">for</span> my $md ( @{$mds} ) {</div>
<div class="line">      $self-&gt;_fetch_children($md);</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $mds;</div>
<div class="line">} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a3a2b1e2023eff92e02ece082bcb2fbc3">_fetch_generic</a></div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a44a14301f9e35e3cf9829c8f42babf78"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_generic_with_args </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : hashref of arguments by column
  Arg        : (optional) if set to 1, all children will be fetched
  Description: Instantiate a <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">GenomeInfo</a> from the database using a 
               generic method, with the supplied arguments
  Returntype : Arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_generic_with_args" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_generic_with_args-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_generic_with_args-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_generic_with_args-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a44a14301f9e35e3cf9829c8f42babf78">_fetch_generic_with_args</a> {</div>
<div class="line">  my ( $self, $args, $keen ) = @_;</div>
<div class="line">  my $sql    = $base_fetch_sql;</div>
<div class="line">  my $params = [ values %$args ];</div>
<div class="line">  my $clause = join( <span class="charliteral">&#39;,&#39;</span>, map { $_ . <span class="stringliteral">&#39;=?&#39;</span> } keys %$args );</div>
<div class="line">  <span class="keywordflow">if</span> ( $clause ne <span class="stringliteral">&#39;&#39;</span> ) {</div>
<div class="line">    $sql .= <span class="stringliteral">&#39; where &#39;</span> . $clause;</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_generic( $sql, $params, $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a788d3fccde1fca75116e4a08576589f2"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_other_alignments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> 
  Description: Add other_alignments to supplied object
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_other_alignments" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_other_alignments-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_other_alignments-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_other_alignments-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a788d3fccde1fca75116e4a08576589f2">_fetch_other_alignments</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  croak</div>
<div class="line"><span class="stringliteral">&quot;Cannot fetch alignments for a GenomeInfo object that has not been stored&quot;</span></div>
<div class="line">    <span class="keywordflow">if</span> !defined $genome-&gt;dbID();</div>
<div class="line">  my $alignments = {};</div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_no_return(</div>
<div class="line">    -SQL =&gt;</div>
<div class="line">      <span class="stringliteral">&#39;select type,name,count from genome_alignment where genome_id=?&#39;</span>,</div>
<div class="line">    -CALLBACK =&gt; sub {</div>
<div class="line">      my @row = @{ shift @_ };</div>
<div class="line">      $alignments-&gt;{ $row[0] }-&gt;{ $row[1] } = $row[2];</div>
<div class="line">      <span class="keywordflow">return</span>;</div>
<div class="line">    },</div>
<div class="line">    -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line">  $genome-&gt;other_alignments($alignments);</div>
<div class="line">  $genome-&gt;has_other_alignments();</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac45ecaf64a7ae889a88cf6bb493ec739"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_publications </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> 
  Description: Add publications to supplied object
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_publications" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_publications-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_publications-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_publications-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac45ecaf64a7ae889a88cf6bb493ec739">_fetch_publications</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  croak</div>
<div class="line"><span class="stringliteral">&quot;Cannot fetch publications for a GenomeInfo object that has not been stored&quot;</span></div>
<div class="line">    <span class="keywordflow">if</span> !defined $genome-&gt;dbID();</div>
<div class="line">  my $pubs =</div>
<div class="line">    $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_simple(</div>
<div class="line">       -SQL =&gt;</div>
<div class="line">         <span class="stringliteral">&#39;select publication from genome_publication where genome_id=?&#39;</span>,</div>
<div class="line">       -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line">  $genome-&gt;publications($pubs);</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a1af91719c30050bf0075ec6d9e3b73bd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_sequences </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> 
  Description: Add sequences to supplied object
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_sequences" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_sequences-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_sequences-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_sequences-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a1af91719c30050bf0075ec6d9e3b73bd">_fetch_sequences</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  croak</div>
<div class="line"><span class="stringliteral">&quot;Cannot fetch sequences for a GenomeInfo object that has not been stored&quot;</span></div>
<div class="line">    <span class="keywordflow">if</span> !defined $genome-&gt;dbID();</div>
<div class="line">  my $sequences =</div>
<div class="line">    $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute(</div>
<div class="line">       -USE_HASHREFS =&gt; 1,</div>
<div class="line">       -SQL =&gt; <span class="stringliteral">&#39;select name,acc from genome_sequence where genome_id=?&#39;</span>,</div>
<div class="line">       -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line">  $genome-&gt;sequences($sequences);</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aa584d1c91e4b551a6cf2e6f8a02d18f3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_fetch_variations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> 
  Description: Add variations to supplied object
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_fetch_variations" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_fetch_variations-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_fetch_variations-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_fetch_variations-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa584d1c91e4b551a6cf2e6f8a02d18f3">_fetch_variations</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  croak</div>
<div class="line"><span class="stringliteral">&quot;Cannot fetch variations for a GenomeInfo object that has not been stored&quot;</span></div>
<div class="line">    <span class="keywordflow">if</span> !defined $genome-&gt;dbID();</div>
<div class="line">  my $variations = {};</div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_no_return(</div>
<div class="line">    -SQL =&gt;</div>
<div class="line">      <span class="stringliteral">&#39;select type,name,count from genome_variation where genome_id=?&#39;</span>,</div>
<div class="line">    -CALLBACK =&gt; sub {</div>
<div class="line">      my @row = @{ shift @_ };</div>
<div class="line">      $variations-&gt;{ $row[0] }-&gt;{ $row[1] } = $row[2];</div>
<div class="line">      <span class="keywordflow">return</span>;</div>
<div class="line">    },</div>
<div class="line">    -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line">  $genome-&gt;variations($variations);</div>
<div class="line">  $genome-&gt;has_variations();</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aa1b982df106fad0f575052e5023a45c2"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_first_element </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : arrayref
  Description: Utility method to return the first element in a list, undef if empty
  Returntype : arrayref element
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_first_element" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_first_element-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_first_element-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_first_element-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a> {</div>
<div class="line">  my ($arr) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $arr &amp;&amp; scalar(@$arr) &gt; 0 ) {</div>
<div class="line">    <span class="keywordflow">return</span> $arr-&gt;[0];</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">else</span> {</div>
<div class="line">    <span class="keywordflow">return</span> undef;</div>
<div class="line">  }</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a1439390262ce0abc462135515ed7922d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected Object Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_get_cached_obj </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : type of object to retrieve
  Arg        : ID of object to retrieve
  Description: Retrieve object from internal cache
  Returntype : object
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_get_cached_obj" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_get_cached_obj-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_get_cached_obj-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_get_cached_obj-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a1439390262ce0abc462135515ed7922d">_get_cached_obj</a> {</div>
<div class="line">  my ( $self, $type, $id ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_cache($type)-&gt;{$id};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="adb913c752b8a9bf2d4c29741634e5e6c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_store_aliases </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Description: Stores the aliases for the supplied object
  Returntype : None
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_store_aliases" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_store_aliases-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_store_aliases-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_store_aliases-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#adb913c752b8a9bf2d4c29741634e5e6c">_store_aliases</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line"></div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">                  -SQL =&gt; q/<span class="keyword">delete</span> from genome_alias where genome_id=?/,</div>
<div class="line">                  -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line"></div>
<div class="line">  <span class="keywordflow">for</span> my $alias ( @{ $genome-&gt;aliases() } ) {</div>
<div class="line">    $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">      -SQL =&gt; q/insert into genome_alias(genome_id,alias)</div>
<div class="line">        values(?,?)/,</div>
<div class="line">      -PARAMS =&gt; [ $genome-&gt;dbID(), $alias ] );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ad80eff92443d9527737e8883b149cd7b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_store_alignments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Description: Stores the alignments for the supplied object
  Returntype : None
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_store_alignments" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_store_alignments-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_store_alignments-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_store_alignments-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ad80eff92443d9527737e8883b149cd7b">_store_alignments</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line"></div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">              -SQL =&gt; q/<span class="keyword">delete</span> from genome_alignment where genome_id=?/,</div>
<div class="line">              -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line"></div>
<div class="line">  <span class="keywordflow">while</span> ( my ( $type, $f ) = each %{ $genome-&gt;other_alignments() } ) {</div>
<div class="line">    <span class="keywordflow">while</span> ( my ( $key, $count ) = each %$f ) {</div>
<div class="line">      $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">        -SQL =&gt;</div>
<div class="line">          q/insert into genome_alignment(genome_id,type,name,count)</div>
<div class="line">        values(?,?,?,?)/,</div>
<div class="line">        -PARAMS =&gt; [ $genome-&gt;dbID(), $type, $key, $count ] );</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a5bbc6aeefb9a713e23c7402019086034"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_store_annotations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Description: Stores the annotations for the supplied object
  Returntype : None
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_store_annotations" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_store_annotations-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_store_annotations-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_store_annotations-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a5bbc6aeefb9a713e23c7402019086034">_store_annotations</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line"></div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">             -SQL =&gt; q/<span class="keyword">delete</span> from genome_annotation where genome_id=?/,</div>
<div class="line">             -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line"></div>
<div class="line">  <span class="keywordflow">while</span> ( my ( $type, $count ) = each %{ $genome-&gt;annotations() } ) {</div>
<div class="line">    $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">      -SQL =&gt; q/insert into genome_annotation(genome_id,type,count)</div>
<div class="line">        values(?,?,?)/,</div>
<div class="line">      -PARAMS =&gt; [ $genome-&gt;dbID(), $type, $count ] );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ae3fc01bba18f5e5ab7a3d34f8849c1c5"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_store_cached_obj </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : type of object to store
  Arg        : object to store
  Description: Store object in internal cache
  Returntype : none
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_store_cached_obj" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_store_cached_obj-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_store_cached_obj-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_store_cached_obj-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ae3fc01bba18f5e5ab7a3d34f8849c1c5">_store_cached_obj</a> {</div>
<div class="line">  my ( $self, $type, $obj ) = @_;</div>
<div class="line">  $self-&gt;_cache($type)-&gt;{ $obj-&gt;dbID() } = $obj;</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a41218a150218370f0c1cdd75ee8812a6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_store_features </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Description: Stores the features for the supplied object
  Returntype : None
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_store_features" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_store_features-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_store_features-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_store_features-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a41218a150218370f0c1cdd75ee8812a6">_store_features</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line"></div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">                -SQL =&gt; q/<span class="keyword">delete</span> from genome_feature where genome_id=?/,</div>
<div class="line">                -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line"></div>
<div class="line">  <span class="keywordflow">while</span> ( my ( $type, $f ) = each %{ $genome-&gt;features() } ) {</div>
<div class="line">    <span class="keywordflow">while</span> ( my ( $analysis, $count ) = each %$f ) {</div>
<div class="line">      $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">        -SQL =&gt;</div>
<div class="line">          q/insert into genome_feature(genome_id,type,analysis,count)</div>
<div class="line">        values(?,?,?,?)/,</div>
<div class="line">        -PARAMS =&gt; [ $genome-&gt;dbID(), $type, $analysis, $count ] );</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a49ed7b29d3e75215103dceaade9ad455"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_store_sequences </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Description: Stores the sequences for the supplied object
  Returntype : None
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_store_sequences" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_store_sequences-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_store_sequences-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_store_sequences-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a49ed7b29d3e75215103dceaade9ad455">_store_sequences</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line"></div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">               -SQL =&gt; q/<span class="keyword">delete</span> from genome_sequence where genome_id=?/,</div>
<div class="line">               -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line"></div>
<div class="line">  my $it = natatime 1000, @{ $genome-&gt;sequences() };</div>
<div class="line">  <span class="keywordflow">while</span> ( my @vals = $it-&gt;() ) {</div>
<div class="line">    my $sql =</div>
<div class="line">      <span class="stringliteral">&#39;insert ignore into genome_sequence(genome_id,name,acc) values &#39;</span></div>
<div class="line">      . join(</div>
<div class="line">      <span class="charliteral">&#39;,&#39;</span>,</div>
<div class="line">      map {</div>
<div class="line">        <span class="charliteral">&#39;(&#39;</span> . $genome-&gt;dbID() . <span class="stringliteral">&#39;,&quot;&#39;</span> . $_-&gt;{name} . <span class="stringliteral">&#39;&quot;,&#39;</span> .</div>
<div class="line">          ( $_-&gt;{acc} ? ( <span class="charliteral">&#39;&quot;&#39;</span> . $_-&gt;{acc} . <span class="charliteral">&#39;&quot;&#39;</span> ) : (<span class="stringliteral">&#39;NULL&#39;</span>) ) . <span class="charliteral">&#39;)&#39;</span></div>
<div class="line">      } @vals );</div>
<div class="line">    $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update( -SQL =&gt; $sql );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a00fc0551c262bce91e305ca859325738"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">protected void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::_store_variations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Description: Stores the variations for the supplied object
  Returntype : None
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-_store_variations" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-_store_variations-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-_store_variations-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-_store_variations-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a00fc0551c262bce91e305ca859325738">_store_variations</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line"></div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">              -SQL =&gt; q/<span class="keyword">delete</span> from genome_variation where genome_id=?/,</div>
<div class="line">              -PARAMS =&gt; [ $genome-&gt;dbID() ] );</div>
<div class="line"></div>
<div class="line">  <span class="keywordflow">while</span> ( my ( $type, $f ) = each %{ $genome-&gt;variations() } ) {</div>
<div class="line">    <span class="keywordflow">while</span> ( my ( $key, $count ) = each %$f ) {</div>
<div class="line">      $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">        -SQL =&gt;</div>
<div class="line">          q/insert into genome_variation(genome_id,type,name,count)</div>
<div class="line">        values(?,?,?,?)/,</div>
<div class="line">        -PARAMS =&gt; [ $genome-&gt;dbID(), $type, $key, $count ] );</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ab92da85659d7ffd46da7bad148d448b4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::build_adaptor </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg  : (optional) Ensembl Genomes release (e.g. 21)
  Example    :</pre><div class="fragment"><div class="line">$info = <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>-&gt;build_adaptor();</div>
</div><!-- fragment --> <pre>  Description: Creates a new adaptor using the Ensembl Genomes public MySQL instance (using the latest release if specified)
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor</a>
  Exceptions : croaks if suitable database cannot be found
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-build_adaptor" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-build_adaptor-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-build_adaptor-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-build_adaptor-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab92da85659d7ffd46da7bad148d448b4">build_adaptor</a> {</div>
<div class="line">  my ( $self, $release ) = @_;</div>
<div class="line">  my $dbname;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $release ) {</div>
<div class="line">    $dbname = PUBLIC_DBNAME . $release;</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">else</span> {</div>
<div class="line"><span class="preprocessor">    # get the latest release</span></div>
<div class="line"><span class="preprocessor"></span>    my $dbc =</div>
<div class="line">      Bio::EnsEMBL::DBSQL::DBConnection-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ada57d4b94bf938786da91eb2b97a754b">new</a>( -user =&gt; eg_user(),</div>
<div class="line">                                              -host =&gt; eg_host(),</div>
<div class="line">                                              -port =&gt; eg_port() );</div>
<div class="line"></div>
<div class="line">    my $dbs = $dbc-&gt;sql_helper()-&gt;execute(</div>
<div class="line">      -SQL =&gt; q/select schema_name, </div>
<div class="line">      cast(replace(schema_name,<span class="stringliteral">&#39;ensemblgenomes_info_&#39;</span>,<span class="stringliteral">&#39;&#39;</span>) as <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> ) as rel</div>
<div class="line">      from information_schema.SCHEMATA </div>
<div class="line">      where schema_name like <span class="stringliteral">&#39;ensemblgenomes_info_%&#39;</span> order by rel desc/</div>
<div class="line">    );</div>
<div class="line">    $dbc-&gt;disconnect_if_idle();</div>
<div class="line"></div>
<div class="line">    <span class="keywordflow">if</span> ( scalar(@$dbs) == 0 ) {</div>
<div class="line">      croak( <span class="stringliteral">&quot;No suitable database found on &quot;</span> .</div>
<div class="line">             eg_host() . <span class="stringliteral">&quot;:&quot;</span> . eg_port() );</div>
<div class="line">    }</div>
<div class="line"></div>
<div class="line">    $dbname = $dbs-&gt;[0][0];</div>
<div class="line"></div>
<div class="line">  }</div>
<div class="line"></div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ada57d4b94bf938786da91eb2b97a754b">new</a>(</div>
<div class="line">                               -DBC =&gt;</div>
<div class="line">                                 Bio::EnsEMBL::DBSQL::DBConnection-&gt;<span class="keyword">new</span>(</div>
<div class="line">                                                     -user =&gt; eg_user(),</div>
<div class="line">                                                     -host =&gt; eg_host(),</div>
<div class="line">                                                     -port =&gt; eg_port(),</div>
<div class="line">                                                     -dbname =&gt; $dbname</div>
<div class="line">                                 ) );</div>
<div class="line">} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab92da85659d7ffd46da7bad148d448b4">build_adaptor</a></div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a98fec113b5f7a90ebfc7473692967325"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch all genome info
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a98fec113b5f7a90ebfc7473692967325">fetch_all</a> {</div>
<div class="line">  my ( $self, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( {}, $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac8ff7541994238568529e5dace97b299"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_by_dbname </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : Name of database
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified database
  Returntype : arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_by_dbname" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_by_dbname-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_by_dbname-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_by_dbname-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac8ff7541994238568529e5dace97b299">fetch_all_by_dbname</a> {</div>
<div class="line">  my ( $self, $name, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;dbname&#39;</span>, $name }, $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a7d9aa9d28b19da134c14ab77a85f6c23"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_by_division </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : Name of division
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome infos for specified division
  Returntype : Arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_by_division" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_by_division-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_by_division-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_by_division-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a7d9aa9d28b19da134c14ab77a85f6c23">fetch_all_by_division</a> {</div>
<div class="line">  my ( $self, $division, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;division&#39;</span>, $division },</div>
<div class="line">                                          $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ab05c482f397d29a8dc3993a8c7c3113d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_by_name_pattern </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : Regular expression matching of genome
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified species
  Returntype : Arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_by_name_pattern" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_by_name_pattern-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_by_name_pattern-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_by_name_pattern-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab05c482f397d29a8dc3993a8c7c3113d">fetch_all_by_name_pattern</a> {</div>
<div class="line">  my ( $self, $name, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    $self-&gt;_fetch_generic(</div>
<div class="line">         $base_fetch_sql . q/ where species REGEXP ? or name REGEXP ? /,</div>
<div class="line">         [ $name, $name ], $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ab3ebe74e0af6996ddd371bd32d9bbe2a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_by_sequence_accession </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : INSDC sequence accession e.g. U00096.1 or U00096
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified sequence accession
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_by_sequence_accession" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_by_sequence_accession-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_by_sequence_accession-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_by_sequence_accession-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab3ebe74e0af6996ddd371bd32d9bbe2a">fetch_all_by_sequence_accession</a> {</div>
<div class="line">  my ( $self, $id, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( $id =~ m/\.[0-9]+$/ ) {</div>
<div class="line">    <span class="keywordflow">return</span> $self-&gt;fetch_all_by_sequence_accession_versioned( $id,</div>
<div class="line">                                                             $keen );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">else</span> {</div>
<div class="line">    <span class="keywordflow">return</span> $self-&gt;fetch_all_by_sequence_accession_unversioned( $id,</div>
<div class="line">                                                               $keen );</div>
<div class="line">  }</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac32ebacac51ec0d2b59db0b573b61c2a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_by_sequence_accession_unversioned </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : INSDC sequence accession e.g. U00096
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified sequence accession
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_by_sequence_accession_unversioned" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_by_sequence_accession_unversioned-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_by_sequence_accession_unversioned-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_by_sequence_accession_unversioned-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac32ebacac51ec0d2b59db0b573b61c2a">fetch_all_by_sequence_accession_unversioned</a> {</div>
<div class="line">  my ( $self, $id, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    $self-&gt;_fetch_generic(</div>
<div class="line">    $base_fetch_sql .</div>
<div class="line"><span class="stringliteral">&#39; where genome_id in (select distinct(genome_id) from genome_sequence where acc like ? or name like ?)&#39;</span>,</div>
<div class="line">    [ $id . <span class="stringliteral">&#39;.%&#39;</span>, $id . <span class="stringliteral">&#39;.%&#39;</span> ],</div>
<div class="line">    $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aa57e37a0374bd7a0de661d4c9f5265a4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_by_sequence_accession_versioned </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : INSDC sequence accession e.g. U00096.1
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified sequence accession
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_by_sequence_accession_versioned" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_by_sequence_accession_versioned-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_by_sequence_accession_versioned-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_by_sequence_accession_versioned-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa57e37a0374bd7a0de661d4c9f5265a4">fetch_all_by_sequence_accession_versioned</a> {</div>
<div class="line">  my ( $self, $id, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    $self-&gt;_fetch_generic(</div>
<div class="line">    $base_fetch_sql .</div>
<div class="line"><span class="stringliteral">&#39; where genome_id in (select distinct(genome_id) from genome_sequence where acc=? or name=?)&#39;</span>,</div>
<div class="line">    [ $id, $id ],</div>
<div class="line">    $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac85701101dbecf4b1f5c58574dcb2f2c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_by_taxonomy_branch </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1TaxonomyNode.html">Bio::EnsEMBL::TaxonomyNode</a>
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified taxonomy node and its children
  Returntype : arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_by_taxonomy_branch" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_by_taxonomy_branch-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_by_taxonomy_branch-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_by_taxonomy_branch-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac85701101dbecf4b1f5c58574dcb2f2c">fetch_all_by_taxonomy_branch</a> {</div>
<div class="line">  my ( $self, $root, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( ref($root) ne <span class="stringliteral">&#39;Bio::EnsEMBL::TaxonomyNode&#39;</span> ) {</div>
<div class="line">    <span class="keywordflow">if</span> ( looks_like_number($root) ) {</div>
<div class="line">      $root = $self-&gt;taxonomy_adaptor()-&gt;fetch_by_taxon_id($root);</div>
<div class="line">    }</div>
<div class="line">    <span class="keywordflow">else</span> {</div>
<div class="line">      ($root) =</div>
<div class="line">        @{ $self-&gt;taxonomy_adaptor()-&gt;fetch_all_by_name($root) };</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line">  my @taxids = ( $root-&gt;taxon_id(),</div>
<div class="line">                 @{ $root-&gt;adaptor()-&gt;fetch_descendant_ids($root) } );</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;fetch_all_by_taxonomy_ids( \@taxids );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="afcb2ec45290da01dea42ee09e9148e7b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_by_taxonomy_id </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-fetch_all_by_taxonomy_id" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_by_taxonomy_id-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_by_taxonomy_id-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_by_taxonomy_id-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#afcb2ec45290da01dea42ee09e9148e7b">fetch_all_by_taxonomy_id</a> {</div>
<div class="line">  my ( $self, $id, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;taxonomy_id&#39;</span>, $id },</div>
<div class="line">                                          $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aff91c234f1f9358536087f5a8ba5b251"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_by_taxonomy_ids </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-fetch_all_by_taxonomy_ids" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_by_taxonomy_ids-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_by_taxonomy_ids-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_by_taxonomy_ids-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aff91c234f1f9358536087f5a8ba5b251">fetch_all_by_taxonomy_ids</a> {</div>
<div class="line">  my ( $self, $ids, $keen ) = @_;</div>
<div class="line"><span class="preprocessor">  # filter list down</span></div>
<div class="line"><span class="preprocessor"></span>  my %ids = map { $_ =&gt; 1 } @$ids;</div>
<div class="line">  my @tids = grep { defined $ids{$_} }</div>
<div class="line">    @{</div>
<div class="line">    $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_simple(</div>
<div class="line">                      -SQL =&gt; q/select distinct taxonomy_id from genome/</div>
<div class="line">    ) };</div>
<div class="line">  my @genomes = ();</div>
<div class="line">  my $it = natatime 1000, @tids;</div>
<div class="line">  <span class="keywordflow">while</span> ( my @vals = $it-&gt;() ) {</div>
<div class="line">    my $sql = $base_fetch_sql . <span class="stringliteral">&#39; where taxonomy_id in (&#39;</span> .</div>
<div class="line">      join( <span class="charliteral">&#39;,&#39;</span>, @vals ) . <span class="charliteral">&#39;)&#39;</span>;</div>
<div class="line">    @genomes = ( @genomes, @{ $self-&gt;_fetch_generic( $sql, [] ) } );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> \@genomes;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a7407979a544da19c04c26f0e3dc4b215"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Array Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_compara_by_division </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : Division of compara analyses to retrieve
  Description: Fetch compara specified compara analysis
  Returntype : array ref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_compara_by_division" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_compara_by_division-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_compara_by_division-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_compara_by_division-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a7407979a544da19c04c26f0e3dc4b215">fetch_all_compara_by_division</a> {</div>
<div class="line">  my ( $self, $division ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_compara_with_args( { division =&gt; $division } );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac4eaa65a71d64e2e90b5a1469e09b8d4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Array Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_compara_by_method </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : Method of compara analyses to retrieve
  Description: Fetch compara specified compara analysis
  Returntype : array ref of  <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_compara_by_method" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_compara_by_method-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_compara_by_method-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_compara_by_method-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac4eaa65a71d64e2e90b5a1469e09b8d4">fetch_all_compara_by_method</a> {</div>
<div class="line">  my ( $self, $method ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_compara_with_args( { method =&gt; $method } );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="af2d425b3fd6e7d0188ead526bad562bf"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Array Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_comparas </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Description: Fetch all compara analyses
  Returntype : array ref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_comparas" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_comparas-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_comparas-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_comparas-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af2d425b3fd6e7d0188ead526bad562bf">fetch_all_comparas</a> {</div>
<div class="line">  my ( $self, $id ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_compara_with_args();</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a0fed890cc2d6b3a5a1452b6873c19d0c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_with_compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch all genome info that have any compara or whole genome alignment data
  Returntype : arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_with_compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_with_compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_with_compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_with_compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a0fed890cc2d6b3a5a1452b6873c19d0c">fetch_all_with_compara</a> {</div>
<div class="line">  my ( $self, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;_fetch_generic( $base_fetch_sql .</div>
<div class="line">q/ where has_genome_alignments=1 or has_pan_compara=1 or has_peptide_compara=1/</div>
<div class="line">  );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aa69042b2065f079af54aa363f916f3e2"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_with_genome_alignments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch all genome info that have whole genome alignment data
  Returntype : arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_with_genome_alignments" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_with_genome_alignments-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_with_genome_alignments-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_with_genome_alignments-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa69042b2065f079af54aa363f916f3e2">fetch_all_with_genome_alignments</a> {</div>
<div class="line">  my ( $self, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_genome_alignments&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> },</div>
<div class="line">                                     $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a0f9ead633abe1821d805c70def8e0f9b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_with_other_alignments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-fetch_all_with_other_alignments" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_with_other_alignments-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_with_other_alignments-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_with_other_alignments-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a0f9ead633abe1821d805c70def8e0f9b">fetch_all_with_other_alignments</a> {</div>
<div class="line">  my ( $self, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_other_alignments&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> },</div>
<div class="line">                                     $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ab8567f2e206b6c653ed1163283c6974e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_with_pan_compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch all genome info that have pan comapra data
  Returntype : arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_with_pan_compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_with_pan_compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_with_pan_compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_with_pan_compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab8567f2e206b6c653ed1163283c6974e">fetch_all_with_pan_compara</a> {</div>
<div class="line">  my ( $self, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_pan_compara&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> },</div>
<div class="line">                                     $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac55295a8f575021917f6e2ac61f4d18c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_with_peptide_compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch all genome info that have peptide compara data
  Returntype : arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_with_peptide_compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_with_peptide_compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_with_peptide_compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_with_peptide_compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac55295a8f575021917f6e2ac61f4d18c">fetch_all_with_peptide_compara</a> {</div>
<div class="line">  my ( $self, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_peptide_compara&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> },</div>
<div class="line">                                     $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a92ec339a777775b5a771791f47189328"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_all_with_variation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch all genome info that have variation data
  Returntype : arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_all_with_variation" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_all_with_variation-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_all_with_variation-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_all_with_variation-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a92ec339a777775b5a771791f47189328">fetch_all_with_variation</a> {</div>
<div class="line">  my ( $self, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_variations&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> },</div>
<div class="line">                                     $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac2dffbf3e25b19e11a9f0432ba42a655"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_by_alias </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : Alias of genome
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified species
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_by_alias" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_by_alias-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_by_alias-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_by_alias-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac2dffbf3e25b19e11a9f0432ba42a655">fetch_by_alias</a> {</div>
<div class="line">  my ( $self, $name, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a>(</div>
<div class="line">              $self-&gt;_fetch_generic(</div>
<div class="line">                $base_fetch_sql .</div>
<div class="line">                  q/ join genome_alias using (genome_id) where alias=?/,</div>
<div class="line">                [$name],</div>
<div class="line">                $keen ) );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a1e18e1443071bedce172e383ce7ab033"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_by_any_name </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-fetch_by_any_name" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_by_any_name-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_by_any_name-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_by_any_name-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a1e18e1443071bedce172e383ce7ab033">fetch_by_any_name</a> {</div>
<div class="line">  my ( $self, $name, $keen ) = @_;</div>
<div class="line">  my $dba = $self-&gt;fetch_by_name( $name, $keen );</div>
<div class="line">  <span class="keywordflow">if</span> ( !defined $dba ) {</div>
<div class="line">    $dba = $self-&gt;fetch_by_species( $name, $keen );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">if</span> ( !defined $dba ) {</div>
<div class="line">    $dba = $self-&gt;fetch_by_alias( $name, $keen );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $dba;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a96453090c2be26ed6664d24f3b09b53d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_by_assembly_id </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : INSDC assembly accession
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified assembly ID (versioned or unversioned)
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_by_assembly_id" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_by_assembly_id-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_by_assembly_id-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_by_assembly_id-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a96453090c2be26ed6664d24f3b09b53d">fetch_by_assembly_id</a> {</div>
<div class="line">  my ( $self, $id, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( $id =~ m/\.[0-9]+$/ ) {</div>
<div class="line">    <span class="keywordflow">return</span> $self-&gt;fetch_by_assembly_id_versioned( $id, $keen );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">else</span> {</div>
<div class="line">    <span class="keywordflow">return</span> $self-&gt;fetch_by_assembly_id_unversioned( $id, $keen );</div>
<div class="line">  }</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ae4e1603ff44789591809234c959bc335"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_by_assembly_id_unversioned </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : INSDC assembly set chain (unversioned accession)
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified assembly set chain
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_by_assembly_id_unversioned" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_by_assembly_id_unversioned-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_by_assembly_id_unversioned-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_by_assembly_id_unversioned-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ae4e1603ff44789591809234c959bc335">fetch_by_assembly_id_unversioned</a> {</div>
<div class="line">  my ( $self, $id, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a>( $self-&gt;_fetch_generic(</div>
<div class="line">                          $base_fetch_sql . <span class="stringliteral">&#39; where assembly_id like ?&#39;</span>,</div>
<div class="line">                          [ $id . <span class="stringliteral">&#39;.%&#39;</span> ], $keen ) );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="af3e0556cf0741089d1c3be618ae01e1c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_by_assembly_id_versioned </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : INSDC assembly accession
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified assembly ID
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_by_assembly_id_versioned" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_by_assembly_id_versioned-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_by_assembly_id_versioned-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_by_assembly_id_versioned-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af3e0556cf0741089d1c3be618ae01e1c">fetch_by_assembly_id_versioned</a> {</div>
<div class="line">  my ( $self, $id, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a>(</div>
<div class="line">     $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;assembly_id&#39;</span>, $id }, $keen ) );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ac2d36ac6d8bd591f0737224575231d27"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_by_dbID </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : ID of genome info
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified ID
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-fetch_by_dbID" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_by_dbID-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_by_dbID-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_by_dbID-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ac2d36ac6d8bd591f0737224575231d27">fetch_by_dbID</a> {</div>
<div class="line">  my ( $self, $id, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span></div>
<div class="line">    <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a>( $self-&gt;_fetch_generic_with_args({ <span class="stringliteral">&#39;genome_id&#39;</span>, $id }</div>
<div class="line">                    ),</div>
<div class="line">                    $keen );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ad8aaa0c16c88e0223e80907e03dbc4de"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_by_dbIDs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : IDs of genome info
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified ID
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-fetch_by_dbIDs" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_by_dbIDs-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_by_dbIDs-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_by_dbIDs-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ad8aaa0c16c88e0223e80907e03dbc4de">fetch_by_dbIDs</a> {</div>
<div class="line">  my ( $self, $ids, $keen ) = @_;</div>
<div class="line">  my @genomes = ();</div>
<div class="line">  my $it = natatime 1000, @{$ids};</div>
<div class="line">  <span class="keywordflow">while</span> ( my @vals = $it-&gt;() ) {</div>
<div class="line">    my $sql = $base_fetch_sql . <span class="stringliteral">&#39; where genome_id in (&#39;</span> .</div>
<div class="line">      join( <span class="charliteral">&#39;,&#39;</span>, @vals ) . <span class="charliteral">&#39;)&#39;</span>;</div>
<div class="line">    @genomes = ( @genomes, @{ $self-&gt;_fetch_generic( $sql, [] ) } );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> \@genomes;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ae544a9ec503f127cd25a3b9bfdf0a6ad"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_by_name </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : Display name of genome 
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified species
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_by_name" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_by_name-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_by_name-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_by_name-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ae544a9ec503f127cd25a3b9bfdf0a6ad">fetch_by_name</a> {</div>
<div class="line">  my ( $self, $name, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a>(</div>
<div class="line">          $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;name&#39;</span>, $name }, $keen ) );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ab05b21c4e09665860c3fb5ea60579cbd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_by_species </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : Name of species
  Arg        : (optional) if 1, expand children of genome info
  Description: Fetch genome info for specified species
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_by_species" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_by_species-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_by_species-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_by_species-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#ab05b21c4e09665860c3fb5ea60579cbd">fetch_by_species</a> {</div>
<div class="line">  my ( $self, $species, $keen ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a>(</div>
<div class="line">    $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;species&#39;</span>, $species }, $keen ) );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aeab1847847a9eece525cd7b8c88a4ddb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_compara_by_dbID </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : ID of compara analysis to retrieve
  Description: Fetch compara specified compara analysis
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_compara_by_dbID" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_compara_by_dbID-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_compara_by_dbID-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_compara_by_dbID-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aeab1847847a9eece525cd7b8c88a4ddb">fetch_compara_by_dbID</a> {</div>
<div class="line">  my ( $self, $id ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a>(</div>
<div class="line">    $self-&gt;_fetch_compara_with_args( { compara_analysis_id =&gt; $id } ) );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a9c54cedfdafaf641a866554bd7e1f70f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a> Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::fetch_compara_by_dbname_method_set </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : DBName of compara analyses to retrieve
  Arg        : Method of compara analyses to retrieve
  Arg        : Set of compara analyses to retrieve
  Description: Fetch specified compara analysis
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-fetch_compara_by_dbname_method_set" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-fetch_compara_by_dbname_method_set-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-fetch_compara_by_dbname_method_set-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-fetch_compara_by_dbname_method_set-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a9c54cedfdafaf641a866554bd7e1f70f">fetch_compara_by_dbname_method_set</a> {</div>
<div class="line">  my ( $self, $dbname, $method, $set_name ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa1b982df106fad0f575052e5023a45c2">_first_element</a>($self-&gt;_fetch_compara_with_args( { dbname =&gt; $dbname, method=&gt;$method, set_name=&gt;$set_name } ));</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a0c4bba091ca39f03f2ebdbc5313118f9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::list_divisions </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Description: Get list of all Ensembl Genomes divisions for which information is available
  Returntype : Arrayref of strings
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-list_divisions" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-list_divisions-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-list_divisions-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-list_divisions-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a0c4bba091ca39f03f2ebdbc5313118f9">list_divisions</a> {</div>
<div class="line">  my ($self) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{dbc}-&gt;sql_helper()</div>
<div class="line">    -&gt;execute_simple( -SQL =&gt;</div>
<div class="line">       q/select distinct division from genome where division&lt;&gt;<span class="stringliteral">&#39;Ensembl&#39;</span>/</div>
<div class="line">    );</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="ada57d4b94bf938786da91eb2b97a754b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::new </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-new" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-new-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-new-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-new-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <span class="keyword">new</span> {</div>
<div class="line">  my ( $proto, @args ) = @_;</div>
<div class="line">  my $self = bless {}, $proto;</div>
<div class="line">  ( $self-&gt;{user}, $self-&gt;{pass}, $self-&gt;{host}, $self-&gt;{port},</div>
<div class="line">    $self-&gt;{dbname}, $self-&gt;{dbc}, $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af138803c65700ed94423cf35c63109cd">taxonomy_adaptor</a>} )</div>
<div class="line">    = rearrange(</div>
<div class="line">                 [ <span class="stringliteral">&#39;USER&#39;</span>,   <span class="stringliteral">&#39;PASS&#39;</span>,</div>
<div class="line">                   <span class="stringliteral">&#39;HOST&#39;</span>,   <span class="stringliteral">&#39;PORT&#39;</span>,</div>
<div class="line">                   <span class="stringliteral">&#39;DBNAME&#39;</span>, <span class="stringliteral">&#39;DBC&#39;</span>,</div>
<div class="line">                   <span class="stringliteral">&#39;TAXONOMY_ADAPTOR&#39;</span> ],</div>
<div class="line">                 @args );</div>
<div class="line">  <span class="keywordflow">if</span> ( !defined $self-&gt;{dbc} ) {</div>
<div class="line">    $self-&gt;{dbc} =</div>
<div class="line">      Bio::EnsEMBL::DBSQL::DBConnection-&gt;new( -USER   =&gt; $self-&gt;{user},</div>
<div class="line">                                              -PASS   =&gt; $self-&gt;{pass},</div>
<div class="line">                                              -HOST   =&gt; $self-&gt;{host},</div>
<div class="line">                                              -PORT   =&gt; $self-&gt;{port},</div>
<div class="line">                                              -DBNAME =&gt; $self-&gt;{dbname}</div>
<div class="line">      );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self;</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a21e83f339f3d250c860f768d68e7e04f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::store </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-store" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-store-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-store-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-store-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a21e83f339f3d250c860f768d68e7e04f">store</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  <span class="keywordflow">return</span> <span class="keywordflow">if</span> defined $genome-&gt;dbID();</div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">    -SQL =&gt;</div>
<div class="line">q/insert into genome(name,species,strain,serotype,division,taxonomy_id,</div>
<div class="line">assembly_id,assembly_name,assembly_level,base_count,</div>
<div class="line">genebuild,dbname,species_id,has_pan_compara,has_variations,has_peptide_compara,</div>
<div class="line">has_genome_alignments,has_synteny,has_other_alignments)</div>
<div class="line">        values(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)/,</div>
<div class="line">    -PARAMS =&gt; [ $genome-&gt;name(),</div>
<div class="line">                 $genome-&gt;species(),</div>
<div class="line">                 $genome-&gt;strain(),</div>
<div class="line">                 $genome-&gt;serotype(),</div>
<div class="line">                 $genome-&gt;division(),</div>
<div class="line">                 $genome-&gt;taxonomy_id(),</div>
<div class="line">                 $genome-&gt;assembly_id(),</div>
<div class="line">                 $genome-&gt;assembly_name(),</div>
<div class="line">                 $genome-&gt;assembly_level(),</div>
<div class="line">                 $genome-&gt;base_count(),</div>
<div class="line">                 $genome-&gt;genebuild(),</div>
<div class="line">                 $genome-&gt;dbname(),</div>
<div class="line">                 $genome-&gt;species_id(),</div>
<div class="line">                 $genome-&gt;has_pan_compara(),</div>
<div class="line">                 $genome-&gt;has_variations(),</div>
<div class="line">                 $genome-&gt;has_peptide_compara(),</div>
<div class="line">                 $genome-&gt;has_genome_alignments(),</div>
<div class="line">                 $genome-&gt;has_synteny(),</div>
<div class="line">                 $genome-&gt;has_other_alignments() ],</div>
<div class="line">    -CALLBACK =&gt; sub {</div>
<div class="line">      my ( $sth, $dbh, $rv ) = @_;</div>
<div class="line">      $genome-&gt;dbID( $dbh-&gt;{mysql_insertid} );</div>
<div class="line">    } );</div>
<div class="line">  $genome-&gt;adaptor($self);</div>
<div class="line">  $self-&gt;_store_aliases($genome);</div>
<div class="line">  $self-&gt;_store_sequences($genome);</div>
<div class="line">  $self-&gt;_store_annotations($genome);</div>
<div class="line">  $self-&gt;_store_features($genome);</div>
<div class="line">  $self-&gt;_store_variations($genome);</div>
<div class="line">  $self-&gt;_store_alignments($genome);</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a21e83f339f3d250c860f768d68e7e04f">store</a></div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="aad8fbb82f254c61dcc029449f5f048ba"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::store_compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Description: Stores the supplied object and all associated  genomes (if not already stored)
  Returntype : None
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-store_compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-store_compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-store_compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-store_compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aad8fbb82f254c61dcc029449f5f048ba">store_compara</a> {</div>
<div class="line">  my ( $self, $compara ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $compara-&gt;dbID() ) {</div>
<div class="line">    <span class="keywordflow">return</span>;</div>
<div class="line">  }</div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">    -SQL =&gt;</div>
<div class="line">      q/insert into compara_analysis(method,division,set_name,dbname)</div>
<div class="line">        values(?,?,?,?)/,</div>
<div class="line">    -PARAMS =&gt; [ $compara-&gt;method(),   $compara-&gt;division(),</div>
<div class="line">                 $compara-&gt;set_name(), $compara-&gt;dbname() ],</div>
<div class="line">    -CALLBACK =&gt; sub {</div>
<div class="line">      my ( $sth, $dbh, $rv ) = @_;</div>
<div class="line">      $compara-&gt;dbID( $dbh-&gt;{mysql_insertid} );</div>
<div class="line">    } );</div>
<div class="line">  <span class="keywordflow">for</span> my $genome ( @{ $compara-&gt;genomes() } ) {</div>
<div class="line">    $self-&gt;store($genome);</div>
<div class="line">    $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">      -SQL =&gt;</div>
<div class="line">q/insert into genome_compara_analysis(genome_id,compara_analysis_id)</div>
<div class="line">        values(?,?)/,</div>
<div class="line">      -PARAMS =&gt; [ $genome-&gt;dbID(), $compara-&gt;dbID() ] );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aad8fbb82f254c61dcc029449f5f048ba">store_compara</a></div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="af138803c65700ed94423cf35c63109cd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::taxonomy_adaptor </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-taxonomy_adaptor" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-taxonomy_adaptor-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-taxonomy_adaptor-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-taxonomy_adaptor-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af138803c65700ed94423cf35c63109cd">taxonomy_adaptor</a> {</div>
<div class="line">  my ( $self, $adaptor ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( defined $adaptor ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af138803c65700ed94423cf35c63109cd">taxonomy_adaptor</a>} = $adaptor;</div>
<div class="line">  }</div>
<div class="line">  elsif ( !defined $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af138803c65700ed94423cf35c63109cd">taxonomy_adaptor</a>} ) {</div>
<div class="line">    $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af138803c65700ed94423cf35c63109cd">taxonomy_adaptor</a>} =</div>
<div class="line">      <a class="code" href="classBio_1_1EnsEMBL_1_1DBSQL_1_1TaxonomyNodeAdaptor.html">Bio::EnsEMBL::DBSQL::TaxonomyNodeAdaptor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1DBSQL_1_1TaxonomyNodeAdaptor.html#a8f07fa12d6dfa71850bdfbc28cb07e2d">new_public</a>();</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span> $self-&gt;{<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#af138803c65700ed94423cf35c63109cd">taxonomy_adaptor</a>};</div>
<div class="line">}</div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a36a75246ac8a7ac25a2c8ac74ae0192e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::update </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeInfo</a>
  Description: Updates the supplied object and all associated child objects (includes other genomes attached by compara if not already stored)
  Returntype : None
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-update" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-update-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-update-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-update-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a36a75246ac8a7ac25a2c8ac74ae0192e">update</a> {</div>
<div class="line">  my ( $self, $genome ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( !defined $genome-&gt;dbID() ) {</div>
<div class="line">    croak <span class="stringliteral">&quot;Cannot update an object that has not already been stored&quot;</span>;</div>
<div class="line">  }</div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">    -SQL =&gt;</div>
<div class="line">q/<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a36a75246ac8a7ac25a2c8ac74ae0192e">update</a> genome <span class="keyword">set</span> name=?,species=?,strain=?,serotype=?,division=?,taxonomy_id=?,</div>
<div class="line">assembly_id=?,assembly_name=?,assembly_level=?,base_count=?,</div>
<div class="line">genebuild=?,dbname=?,species_id=?,has_pan_compara=?,has_variations=?,has_peptide_compara=?,</div>
<div class="line">has_genome_alignments=?,has_synteny=?,has_other_alignments=? where genome_id=?/,</div>
<div class="line">    -PARAMS =&gt; [ $genome-&gt;name(),</div>
<div class="line">                 $genome-&gt;species(),</div>
<div class="line">                 $genome-&gt;strain(),</div>
<div class="line">                 $genome-&gt;serotype(),</div>
<div class="line">                 $genome-&gt;division(),</div>
<div class="line">                 $genome-&gt;taxonomy_id(),</div>
<div class="line">                 $genome-&gt;assembly_id(),</div>
<div class="line">                 $genome-&gt;assembly_name(),</div>
<div class="line">                 $genome-&gt;assembly_level(),</div>
<div class="line">                 $genome-&gt;base_count(),</div>
<div class="line">                 $genome-&gt;genebuild(),</div>
<div class="line">                 $genome-&gt;dbname(),</div>
<div class="line">                 $genome-&gt;species_id(),</div>
<div class="line">                 $genome-&gt;has_pan_compara(),</div>
<div class="line">                 $genome-&gt;has_variations(),</div>
<div class="line">                 $genome-&gt;has_peptide_compara(),</div>
<div class="line">                 $genome-&gt;has_genome_alignments(),</div>
<div class="line">                 $genome-&gt;has_synteny(),</div>
<div class="line">                 $genome-&gt;has_other_alignments(),</div>
<div class="line">                 $genome-&gt;dbID() ] );</div>
<div class="line">  $genome-&gt;adaptor($self);</div>
<div class="line">  $self-&gt;_store_aliases($genome);</div>
<div class="line">  $self-&gt;_store_sequences($genome);</div>
<div class="line">  $self-&gt;_store_annotations($genome);</div>
<div class="line">  $self-&gt;_store_features($genome);</div>
<div class="line">  $self-&gt;_store_variations($genome);</div>
<div class="line">  $self-&gt;_store_alignments($genome);</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a36a75246ac8a7ac25a2c8ac74ae0192e">update</a></div>
</div><!-- fragment --> </div> 
</div>
</div>
<a class="anchor" id="a2b9a46d26f3bb8b3395888d2a5dedf23"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public void Bio::EnsEMBL::Utils::MetaData::DBSQL::GenomeInfoAdaptor::update_compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre>  Arg       : <a class="el" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::Utils::MetaData::GenomeComparaInfo</a>
  Description: Updates the supplied object and all associated  genomes
  Returntype : None
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-update_compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
    <img id='codesection-update_compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-update_compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-update_compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><div class="line">sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a2b9a46d26f3bb8b3395888d2a5dedf23">update_compara</a> {</div>
<div class="line">  my ( $self, $compara ) = @_;</div>
<div class="line">  <span class="keywordflow">if</span> ( !defined $compara-&gt;dbID() ) {</div>
<div class="line">    croak <span class="stringliteral">&quot;Cannot update compara object with no dbID&quot;</span>;</div>
<div class="line">  }</div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">    -SQL =&gt; q/<a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a36a75246ac8a7ac25a2c8ac74ae0192e">update</a> compara_analysis </div>
<div class="line">      <span class="keyword">set</span> method=?,division=?,set_name=?,dbname=? </div>
<div class="line">      where compara_analysis_id=?/,</div>
<div class="line">    -PARAMS =&gt; [ $compara-&gt;method(),   $compara-&gt;division(),</div>
<div class="line">                 $compara-&gt;set_name(), $compara-&gt;dbname(),</div>
<div class="line">                 $compara-&gt;dbID() ] );</div>
<div class="line"></div>
<div class="line">  $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">         -SQL =&gt;</div>
<div class="line">           q/<span class="keyword">delete</span> from genome_compara_analysis where compara_analysis_id=?/,</div>
<div class="line">         -PARAMS =&gt; [ $compara-&gt;dbID() ] );</div>
<div class="line">  <span class="keywordflow">for</span> my $genome ( @{ $compara-&gt;genomes() } ) {</div>
<div class="line">    <span class="keywordflow">if</span> ( defined $genome-&gt;dbID() ) {</div>
<div class="line">      $self-&gt;update($genome);</div>
<div class="line">    }</div>
<div class="line">    <span class="keywordflow">else</span> {</div>
<div class="line">      $self-&gt;store($genome);</div>
<div class="line">    }</div>
<div class="line">    $self-&gt;{dbc}-&gt;sql_helper()-&gt;execute_update(</div>
<div class="line">      -SQL =&gt;</div>
<div class="line">q/insert into genome_compara_analysis(genome_id,compara_analysis_id)</div>
<div class="line">        values(?,?)/,</div>
<div class="line">      -PARAMS =&gt; [ $genome-&gt;dbID(), $compara-&gt;dbID() ] );</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">return</span>;</div>
<div class="line">} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a2b9a46d26f3bb8b3395888d2a5dedf23">update_compara</a></div>
</div><!-- fragment --> </div> 
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>modules/Bio/EnsEMBL/Utils/MetaData/DBSQL/GenomeInfoAdaptor.pm</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.3.1
</small></address>
</body>
</html>
